// @qb/qs-icccd/src/router/index.js
import Vue from 'vue';
import Router from 'vue-router';

Vue.use(Router);

/* Layout */
import Layout from '@icccdPath/pages/layout';

/* Router Modules */
/** note: Submenu only appear when children.length>=1
 *  detail see  https://panjiachen.github.io/vue-element-admin-site/guide/essentials/router-and-nav.html
 **/

/**
 * hidden: true                   if `hidden:true` will not show in the sidebar(default is false)
 * alwaysShow: true               if set true, will always show the root menu, whatever its child routes length
 *                                if not set alwaysShow, only more than one route under the children
 *                                it will becomes nested mode, otherwise not show the root menu
 * redirect: noredirect           if `redirect:noredirect` will no redirect in the breadcrumb
 * name:'router-name'             the name is used by <keep-alive> (must set!!!)
 * meta : {
    permission: ['menu_key1',"menu_key2"]    will control the page permission (you can set multiple permissions)
    title: 'title'               the name show in submenu and breadcrumb (recommend set)
    icon: 'svg-name'             the icon show in the sidebar
    noCache: true                if true, the page will no be cached(default is false)
    breadcrumb: false            if false, the item will hidden in breadcrumb(default is true)
  }
 **/

export const constantRouterMap = [
  {
    path: '/redirect',
    component: Layout,
    hidden: true,
    children: [
      {
        path: '/redirect/:path*',
        component: () => import('@icccdPath/pages/common/redirect/index')
      }
    ]
  },
  {
    path: '/login',
    component: () => import('@icccdPath/pages/common/login/index'),
    hidden: true
  },
  {
    path: '/',
    component: Layout,
    redirect:()=>{
      let path=`${window.systemParams.REDIRECT_PATH?window.systemParams.REDIRECT_PATH:'/dashboard'}`;
      return  path
    },
    meta: { title: 'demo', icon: 'icon-file', noCache: true },
    children: [
      {
        path: 'dashboard',
        component: () => import('@icccdPath/pages/dashboard/index'),
        name: 'dashboard',
        meta: { title: '首页', icon: 'icon-file', noCache: true }
      },
      {
        path: 'scenes-add',
        component: () => import('@icccdPath/pages/scenes/add'),
        name: 'scenes-add',
        meta: { title: '添加场景', icon: 'iconf-Alarm', hidden: true }
      }
    ]
  },
  {
    path: '/themeSimple',
    component: () => import('@icccdPath/pages/dashboard/theme-simple/index'),
    name: 'themeSimple'
  },
  {
    path: '/themeSelfdom',
    component: () => import('@icccdPath/pages/dashboard/theme-selfdom/index'),
    name: 'themeSelfdom'
  },
  // 错误界面
  {
    path: '/403-1',
    component: () => import('@icccdPath/pages/common/error-page/403-1.vue'),
    name: '403-1'
  },
  // 错误界面
  {
    path: '/403-2',
    component: () => import('@icccdPath/pages/common/error-page/403-2.vue'),
    name: '403-2'
  },
  // 错误界面
  {
    path: '/404',
    component: () => import('@icccdPath/pages/common/error-page/404.vue'),
    name: '404'
  },
  // 错误界面
  {
    path: '/500',
    component: () => import('@icccdPath/pages/common/error-page/500.vue'),
    name: '500'
  },
  // 错误界面
  {
    path: '/501',
    component: () => import('@icccdPath/pages/common/error-page/501.vue'),
    name: '501'
  }
];

export default new Router({
  // mode: 'history', // require service support
  scrollBehavior: () => ({ y: 0 }),
  routes: constantRouterMap
});

export const asyncRouterMap = [
  /** When your routing table is too long, you can split it into small modules**/
  { path: '*', redirect: '/404', hidden: true }
];
