// ui-component-v4/lib/dg-directives/clickoutside.js
"use strict";

exports.__esModule = true;

var _vue = require("vue");

var _vue2 = _interopRequireDefault(_vue);

var _dom = require("main/utils/dom");

var _lodash = require("lodash");

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var nodeList = [];
var ctx = "@@clickoutsideContext";

var startClick = void 0;
var seed = 0;

!_vue2.default.prototype.$isServer && (0, _dom.on)(document, "mousedown", function (e) {
  return startClick = e;
});

!_vue2.default.prototype.$isServer && (0, _dom.on)(document, "mouseup", function (e) {
  nodeList.forEach(function (node) {
    return node[ctx].documentHandler(e, startClick);
  });
});
function deepContainPopper(vnode, mousedown, mouseup) {
  // 过滤掉子节点不存在日期和下拉类似浮层点击
  var isContain = false;
  var isFilter = function isFilter(vnode) {
    // 节点实例的存在浮层，且包含点击元素
    if (vnode.context && vnode.context.popperElm && vnode.context.popperElm.contains(mousedown.target)) {
      isContain = true;
      return;
    }
    // 普通节点找子节点，组件占位节点，找第一个根节点
    var children = !_lodash2.default.isEmpty(vnode.children) && vnode.children || !_lodash2.default.isEmpty(vnode.componentInstance) && [vnode.componentInstance._vnode] || [];
    // 递归调用子节点
    children.forEach(isFilter);
  };
  isFilter(vnode);
  return isContain;
}
function createDocumentHandler(el, binding, vnode) {
  return function () {
    var mouseup = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var mousedown = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    if (!vnode || !vnode.context || !mouseup.target || !mousedown.target || el.contains(mouseup.target) || el.contains(mousedown.target) || el === mouseup.target || vnode.context.popperElm && (vnode.context.popperElm.contains(mouseup.target) || vnode.context.popperElm.contains(mousedown.target)) || deepContainPopper(vnode, mousedown, mouseup) // 增加yangjy ，深度遍历子节点是否存在弹层包含点击元素
    ) return;

    if (binding.expression && el[ctx].methodName && vnode.context[el[ctx].methodName]) {
      vnode.context[el[ctx].methodName]();
    } else {
      el[ctx].bindingFn && el[ctx].bindingFn();
    }
  };
}

/**
 * v-clickoutside
 * @desc 点击元素外面才会触发的事件
 * @example
 * ```vue
 * <div v-element-clickoutside="handleClose">
 * ```
 */
exports.default = {
  bind: function bind(el, binding, vnode) {
    nodeList.push(el);
    var id = seed++;
    el[ctx] = {
      id: id,
      documentHandler: createDocumentHandler(el, binding, vnode),
      methodName: binding.expression,
      bindingFn: binding.value
    };
  },
  update: function update(el, binding, vnode) {
    el[ctx].documentHandler = createDocumentHandler(el, binding, vnode);
    el[ctx].methodName = binding.expression;
    el[ctx].bindingFn = binding.value;
  },
  unbind: function unbind(el) {
    var len = nodeList.length;

    for (var i = 0; i < len; i++) {
      if (nodeList[i][ctx].id === el[ctx].id) {
        nodeList.splice(i, 1);
        break;
      }
    }
    delete el[ctx];
  }
};
