// ui-jz-v4-common/src/api/qb-base/dubbo/DubboAttachmentApi.js
import { AbsAttachmentApi } from "../abstract/AbsAttachmentApi";
import qs from "qs";
class DubboAttachmentApi extends AbsAttachmentApi {
    loadFiles(type, ids,loading) {
        return this.request({
            url: `/fjUploader/loadFiles`,
            method: "post",
            loading: loading,
            data: [type, ids],
            baseURL: this.modulePath,
            headers:{'Content-Type':'application/json'}
        });
    }
    delFiles(type, ids,loading) {
        let params = {};
        params.type = type;
        params.ids = ids;
        return this.request({
            url: `/v2/fjUploader/deletFiles`,
            method: "post",
            loading: loading,
            data: qs.stringify(params),
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8" }
        });
    }
    saveFiles(type, ids,loading) {
        let params = {};
        params.type = type;
        params.ids =(typeof ids ==='object') ? JSON.stringify(ids) : ids;
        return this.request({
            url: `/v2/fjUploader/saveFiles`,
            method: "post",
            loading: loading,
            data: qs.stringify(params),
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8" }
        });
    }
    getDatabyCardid(cardid,loading) {
        return this.request({
            url: `/zpfw/showZp?objId=${cardid}`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        });
    }
}
export { DubboAttachmentApi };
