// ui-jz-v4-common/src/api/qb-base/dubbo/DubboBufferedCodedicApi.js
import { AbsBufferedCodedicApi } from "../abstract/AbsBufferedCodedicApi";
//request
class DubboBufferedCodedicApi extends AbsBufferedCodedicApi {
    fetchUrl(url) {
        return this.request({
            url: url,
            method: "get",
            baseURL: this.modulePath
        });
    }
    getCodeMap(codeId, loading, callback) {
        // getBMTableList  获取list结构的
        // getBMTableString 获取字符串的
        this.request({
            url: `/bufferedBmTableService/getBMTableList`,
            method: "post",
            data: [codeId],
            loading: loading,
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/json;charset=UTF-8" }
        }).then(({ data }) => {
            if(data.status=="1"){
                callback(data.data);
            }
            console.info(`加载dubbo codedic:${this.module}, ${codeId}成功...`);
        });
    }
    refreshAll(loading,callback) {
        return this.request({
            url: `/bufferedBmTableService/refreshAll`,
            method: "post",
            loading: loading,
            data: [],
            baseURL: this.modulePath
        }).then(({ data }) => {
            let codeMap = data.data;
            callback(codeMap);
        });
    }
    refreshCodeId(codeId, loading, callback) {
        return this.request({
            url: `/bufferedBmTableService/refreshCodeId`,
            method: "post",
            loading: loading,
            data: [codeId],
            baseURL: this.modulePath
        }).then(({ data }) => {
            let codeMap = data.data;
            callback(codeMap);
        });
    }
}

export { DubboBufferedCodedicApi };
