// ui-jz-v4-common/src/api/qb-base/sboot/SbootBufferedCodedicApi.js
import { AbsBufferedCodedicApi } from "../abstract/AbsBufferedCodedicApi";

class SbootBufferedCodedicApi extends AbsBufferedCodedicApi {
    fetchUrl(url) {
        return this.request({
            url: url,
            method: "get",
            baseURL: this.modulePath
        });
    }
    getCodeMap(codeId, loading,callback) {
        this.request({
            url: `/codeDic/${codeId}?action=loadList`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        }).then(({ data }) => {
            if(data.status=="1"){
                callback(data.data);
            }
            console.info(`加载dubbo codedic:${this.module}, ${codeId}成功...`);
        });
    }
    refreshAll(loading,callback) {
        this.request({
            url: `/codeDic?action=refreshAll`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        }).then(({ data }) => {
            let codeMap = data.data;
            callback(codeMap);
        });
    }
    refreshCodeId(codeId, loading, callback) {
        this.request({
            url: `/codeDic/${codeId}?action=refreshCodeId`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        }).then(({ data }) => {
            let codeMap = data.data;
            callback(codeMap);
        });
    }
}

export { SbootBufferedCodedicApi };
