// ui-jz-v4-common/src/api/qb-base/sboot/SbootChooseGroup.js
import { AbsChooseGroup } from "../abstract/AbsChooseGroup";

class SbootChooseGroup extends AbsChooseGroup {
    /**
     * 获取用户分组
     */
    getUserGroup(codeId,shareCode,loading){
       return this.request({
            url: `/choose/group/v1/group-select/groups/${codeId}/${shareCode}`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        });
    }
    /**
     * 保存用户分组
     */
    saveUserGroup(params,loading){
       return this.request({
            url: "/choose/group/v1/group-select/groups",
            method: "post",
            data: params,
            loading: loading,
            baseURL: this.modulePath,
            headers: {
                "Content-Type": "application/json"
            }
        });
    }
    /**
     * 删除用户分组
     */
    delUserGroup(groupId,loading){
        return this.request({
            url: `choose/group/v1/group-select/groups/${groupId}`,
            method: "DELETE",
            loading: true,
            baseURL: this.modulePath
        });
    }
    /**
     * 获取常用分组
     */
    getMostGroup(codeId,userId,loading){
        return this.request({
            url: `/choose/code/v1/group-select/most/${codeId}/${userId}`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        });
    }
    /**
     * 保存常用分组
     */
    saveMostGroup(params,loading){
       return this.request({
            url: "/choose/code/v1/group-select/most",
            method: "post",
            data: params,
            loading: loading,
            baseURL: this.modulePath,
            headers: {
                "Content-Type": "application/json"
            }
        });
    }
}

export { SbootChooseGroup };
