// ui-jz-v4-common/src/api/qb-base/sboot/SbootAttachmentApi.js
import { AbsAttachmentApi } from "../abstract/AbsAttachmentApi";
import qs from "qs";
class SbootAttachmentApi extends AbsAttachmentApi {

    loadFiles(type, ids,loading) {
        let params = {};
        params.type = type;
        params.ids = JSON.stringify(ids);
        return this.request({
            url: `/fjUploader/loadFiles`,
            method: "post",
            loading: loading,
            data: qs.stringify(params),
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8" }
        });
    }
    delFiles(type, ids,loading) {
        let params = {};
        params.type = type;
        params.ids = JSON.stringify(ids);
        return this.request({
            url: `/fjUploader/deletFiles`,
            method: "post",
            loading: loading,
            data: qs.stringify(params),
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8" }
        });
    }
    saveFiles(type, ids,loading) {
        let params = {};
        params.type = type;
        params.ids = JSON.stringify(ids);
        return this.request({
            url: `/fjUploader/saveFiles`,
            method: "post",
            loading: loading,
            data: qs.stringify(params),
            baseURL: this.modulePath,
            headers: { "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8" }
        });
    }
    getDatabyCardid(cardid,loading) {
        return this.request({
            url: `/zpfw/showZp?objId=${cardid}`,
            method: "GET",
            loading: loading,
            baseURL: this.modulePath
        });
    }
}

export { SbootAttachmentApi };
