// ui-jz-v4-common/src/api/qb-base/AttrachmentApi.js
import { AbsAttachmentApi } from "./abstract/AbsAttachmentApi";
import { DubboAttachmentApi } from "./dubbo/DubboAttachmentApi";
import { SbootAttachmentApi } from "./sboot/SbootAttachmentApi";
/**
 * 附件管理
 */
class AttrachmentApi extends AbsAttachmentApi {
    api = null;
    constructor(module) {
        super(module);
        //1.初始化api
        if (this.moduletype == "sboot") {
            this.api = new SbootAttachmentApi(module);
        } else if (this.moduletype == "dubbo") {
            this.api = new DubboAttachmentApi(module);
        }
    }
    delFiles(type, ids,loadingShow) {
        return this.api.delFiles(type, ids,loadingShow);
    }
    saveFiles(type, ids,loadingShow) {
        return this.api.saveFiles(type, ids,loadingShow);
    }
    loadFiles(type, ids,loadingShow) {
        return this.api.loadFiles(type, ids,loadingShow);
    }
    getDatabyCardid(cardid,loadingShow) {
        return this.api.getDatabyCardid(cardid,loadingShow);
    }
}
export { AttrachmentApi };
