// ui-jz-v4-common/src/api/qb-base/ChooseGroupApi.js
import { AbsChooseGroup } from "./abstract/AbsChooseGroup";
import { DubboChooseGroup } from "./dubbo/DubboChooseGroup";
import { SbootChooseGroup } from "./sboot/SbootChooseGroup";

class ChooseGroupApi extends AbsChooseGroup {
    api = null;
    constructor(module) {
        super(module);
        // 初始化api
        if (this.moduletype === "sboot") {
            this.api = new SbootChooseGroup(module);
        } else if (this.moduletype === "dubbo") {
            this.api = new DubboChooseGroup(module);
        } else {
            throw new Error(`module：${this.module}，无法匹配api错误`);
        }
    }
    /**
     * 获取用户分组
     * codeId 表码ID
     * shareCode 共享代码
     */
    getUserGroup(codeId,shareCode,loading){
        return this.api.getUserGroup(codeId,shareCode,loading);
    }
    /**
     * 保存用户分组
     * 保存参数
     */
    saveUserGroup(params,loading){
        return this.api.saveUserGroup(params,loading);
    }
    /**
     * 删除用户分组
     * groupId 用户分组id
     */
    delUserGroup(groupId,loading){
        return this.api.delUserGroup(groupId,loading);
    }
    /**
     * 获取常用分组
     *表码ID
     * 用户ID
     */
    getMostGroup(codeId,userId,loading){
        return this.api.getMostGroup(codeId,userId,loading);
    }
    /**
     * 保存常用分组
     * params 参数
     */
    saveMostGroup(params,loading){
        return this.api.saveMostGroup(params,loading);
    }
}
export { ChooseGroupApi };
