// @qb/qs-icccd/src/api/gztApi.js
/**
 * 首页工作台
 * create on 2019-11-18 10:02:37
 */
import request from '@icccdPath/utils/request';
import qs from 'qs';

let api = {
  /**
   * 加载用户管理详情
   */
  loadUserManagerDetail: function() {
    return request({
      url: `/gzt/loadUserManagerDetail`,
      method: 'get',
      loading: false
    });
  },
  /**
   * 通过应用编号删除我的应用
   * @param yybh    String
   *        应用编号
   */
  deleteMyAppByYybh: function(yybh) {
    return request({
      url: `/gzt/deleteMyAppByYybh/${yybh}`,
      method: 'get',
      loading: false
    });
  },
  /**
   * 加载我的应用
   */
  loadMyApp: function() {
    return request({
      url: `/gzt/loadMyApp`,
      method: 'get',
      loading: false
    });
  },
  /**
   * 加载用户扩展信息
   */
  loadUserExtInfo: function() {
    return request.get(`/gzt/loadUserExtInfo`, {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
      }
    });
  },
  /**
   * 保存我的所有应用
   * @param userAppVoList    List<UserAppVo>
   *        userAppVoList
   */
  saveMyAllApp: function(userAppVoList) {
    return request({
      url: `/gzt/saveMyAllApp`,
      method: 'post',
      data: userAppVoList,
      loading: false
    });
  },
  /**
   * 保存系统主题和背景图片
   * @param xtzt    String
   *        系统主题(JYFG:简约风格，GXFG：个性风格)
   * @param bjtpid    String
   *        背景图片id
   */
  saveXtztAndBjtp: function(submitData) {
    return request({
      url: `/gzt/saveXtztAndBjtp`,
      method: 'post',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
      },
      data: qs.stringify(submitData),
      loading: false
    });
  },
  /**
   * 背景图片导入
   * @param file    MultipartFile
   *        file
   */
  bjtpImport: function(file) {
    // 组装请求参数，并返回请求实例
    // let submitData = {};
    // submitData.file = (typeof file ==='object') ? JSON.stringify(file) : file;
    // return axios.post(`${mp}/gzt/bjtpImport`, qs.stringify(submitData), {
    //   headers: {
    //     'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
    //   }
    // });
    return request({
      url: `/gzt/bjtpImport`,
      method: 'post',
      data: file,
      headers: {
        // Authorization: TOKEN,
        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
      },
      responseType: 'blob',
      loading: false
    });
  },
  /**
   * 加载最新的系统消息（默认两条）
   */
  loadLatestXtxx: function() {
    return request({
      url: `/gzt/loadLatestXtxx`,
      method: 'get',
      loading: false
    });
  },
  /**
   * 温馨提示-加载随机的常见问题（默认两条）
   */
  loadRandomCjwt: function() {
    return request({
      url: `/gzt/loadRandomCjwt`,
      method: 'get',
      loading: false
    });
  },
  /**
   * 删除背景图片
   * @param bjtpid    String
   *        背景图片id
   */
  deteleBjtpById: function(bjtpid) {
    return request({
      url: `/gzt/deteleBjtpById/${bjtpid}`,
      method: 'DELETE',
      loading: false
    });
  },
  /**
   * 添加我的外部应用
   * @param vo    UserExternalAppVo
   *        vo
   */
  addMyExternalApp: function(vo) {
    return request({
      url: `/gzt/addMyExternalApp`,
      method: 'post',
      data: vo,
      loading: false
    });
  },
  /**
   * 统计背景审查总数
   */
  statBjsczs: function() {
    return request({
      url: `/gzt/statBjsczs`,
      method: 'get',
      loading: false
    });
  },

  /**
   * 统计待办工作总数
   */
  statDbgzzs: function() {
    return request({
      url: `/gzt/statDbgzzs`,
      method: 'get',
      loading: false
    });
  },

  /**
   * 首页待办-加载最新的待办工作（默认两条）
   */
  loadLatestDbgz: function() {
    return request({
      url: `/gzt/loadLatestDbgz`,
      method: 'get',
      loading: false
    });
  }
};
export default api;
