// ui-component-v4/src/index.js
/* Automatically generated by './build/bin/build-entry.js' */

import Scrollbar from '../packages/scrollbar/index.js';
import RadioGroup from '../packages/radio-group/index.js';
import CheckboxGroup from '../packages/checkbox-group/index.js';
import Select from '../packages/select/index.js';
import Transfer from '../packages/transfer/index.js';
import Tree from '../packages/tree/index.js';
import Cascader from '../packages/cascader/index.js';
import Table from '../packages/table/index.js';
import TableColumn from '../packages/table-column/index.js';
import FormTable from '../packages/form-table/index.js';
import FormColumn from '../packages/form-column/index.js';
import Upload from '../packages/upload/index.js';
import Viewer from '../packages/viewer/index.js';
import TreeSelect from '../packages/tree-select/index.js';
import TreeDrop from '../packages/tree-drop/index.js';
import FormItem from '../packages/form-item/index.js';
import Alert from '../packages/alert/index.js';
import Tag from '../packages/tag/index.js';
import Button from '../packages/button/index.js';
import Dialog from '../packages/dialog/index.js';
import MessageBox from '../packages/message-box/index.js';
import TableSelect from '../packages/table-select/index.js';
import Radio from '../packages/radio/index.js';
import DatePicker from '../packages/date-picker/index.js';
import TimePicker from '../packages/time-picker/index.js';
import LogSearch from '../packages/log-search/index.js';
import TimeLine from '../packages/time-line/index.js';
import LineTable from '../packages/line-table/index.js';
import LineColumn from '../packages/line-column/index.js';
import PageCode from '../packages/page-code/index.js';
import Switch from '../packages/switch/index.js';
import Badge from '../packages/badge/index.js';
import Card from '../packages/card/index.js';
import Calendar from '../packages/calendar/index.js';
import Row from '../packages/row/index.js';
import Col from '../packages/col/index.js';
import GroupSelect from '../packages/group-select/index.js';
import Input from '../packages/input/index.js';
import InputNumber from '../packages/input-number/index.js';

const components = [
  Scrollbar,
  RadioGroup,
  CheckboxGroup,
  Select,
  Transfer,
  Tree,
  Cascader,
  Table,
  TableColumn,
  FormTable,
  FormColumn,
  Upload,
  Viewer,
  TreeSelect,
  TreeDrop,
  FormItem,
  Alert,
  Tag,
  Button,
  Dialog,
  TableSelect,
  Radio,
  DatePicker,
  TimePicker,
  LogSearch,
  TimeLine,
  LineTable,
  LineColumn,
  PageCode,
  Switch,
  Badge,
  Card,
  Calendar,
  Row,
  Col,
  GroupSelect,
  Input,
  InputNumber
];

const install = function(Vue, opts = {}) {
  components.forEach(component => {
    Vue.component(component.name, component);
  });

  // +lutz+
  Vue.prototype.$dgMsgbox = MessageBox;
  Vue.prototype.$dgAlert = MessageBox.alert;
  Vue.prototype.$dgConfirm = MessageBox.confirm;
  Vue.prototype.$dgPrompt = MessageBox.prompt;
  // -lutz-
  // +tangdm+
  Vue.prototype.$dgDialog = Dialog.dialog;
  Vue.prototype.$dgLayer = Dialog.layerDialog(Vue);
  Vue.prototype.$layer = Dialog.layer;
  // -tangdm-
};

/* istanbul ignore if */
if (typeof window !== 'undefined' && window.Vue) {
  install(window.Vue);
}

export default {
  version: '1.2.36',
  install,
  Scrollbar,
  RadioGroup,
  CheckboxGroup,
  Select,
  Transfer,
  Tree,
  Cascader,
  Table,
  TableColumn,
  FormTable,
  FormColumn,
  Upload,
  Viewer,
  TreeSelect,
  TreeDrop,
  FormItem,
  Alert,
  Tag,
  Button,
  Dialog,
  MessageBox,
  TableSelect,
  Radio,
  DatePicker,
  TimePicker,
  LogSearch,
  TimeLine,
  LineTable,
  LineColumn,
  PageCode,
  Switch,
  Badge,
  Card,
  Calendar,
  Row,
  Col,
  GroupSelect,
  Input,
  InputNumber
};
