// ui-jz-v4-common/src/components/qb-base/cascader.js
/**
 * 级联选择框(未分页的)
 * code是有层级的表码
 * @Author:
 * @Date:   2019-11-20
 * @Project qb
 */
import { Cascader } from "ui-component-v4";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api
export default {
    name: 'Qb'+Cascader.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        //不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        options: {
            type: Array,
            default: () => [],
            require: false
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },

    mixins: [Cascader],

    methods: {
        //加载数据
        loadData(params) {
            //树形数据加载
            let codedicApi=new BufferedCodedicApi(params.module);
            codedicApi.reqData(params, data => {
                this.$emit('loadEnd',data);
                this.dataSource = data;
            });
        },
        //初始化数据
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
