// ui-jz-v4-common/src/components/qb-base/checkbox-group.js
/**
 * 多选框组
 * @Author:
 * @Date:   2019-02-27
 * @Project qb
 */
import { CheckboxGroup } from "ui-component-v4";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api

export default {
    name: 'QbCheckboxGroup',

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // sql
        sql: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        //不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        customData:{
            type:Array,
            default: () => []
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },
    data() {
        return {
            dataSourceMap: {}
        };
    },
    mixins: [CheckboxGroup],

    methods: {
        getLabel() {
            return this.value.split(",")
                .map(item=>this.dataSourceMap[item]).join(",");
        },
        loadData(params) {
            let codedicApi=new BufferedCodedicApi(params.module);
            codedicApi.reqData(params, data => {
                let result= this.customData.concat(data);
                this.$emit('loadEnd',result);
                this.dataSource = result;
                result.map(item => {
                    this.dataSourceMap[item.value]=item.label;
                })
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
