// ui-jz-v4-common/src/components/qb-base/radio-group.js
/**
 * 单选框组
 * @Author:
 * @Date:   2019-02-25
 * @Project qb
 */
import { RadioGroup } from "ui-component-v4";
//import DataOp from "../../utils/zt-data";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api

export default {
    name: 'QbRadioGroup',

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        //不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        customData:{
            type:Array,
            default: () => []
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },

    mixins: [RadioGroup],

    methods: {
        /**
         * 数据加载的方法
         */
        loadData(params) {
            let codedicApi=new BufferedCodedicApi(params.module);
            codedicApi.reqData(params, data => {
                let result= this.customData.concat(data);
                this.$emit('loadEnd',result);
                this.dataSource = result;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
