// ui-jz-v4-common/src/components/qb-base/select.js
/**
 * 单选下拉框
 * @Author:
 * @Date:   2019-02-27
 * @Project qb
 */
import { Select } from "ui-component-v4";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api

export default {
    name: 'QbSelect',

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // sql
        sql: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        //不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        onLoad: {
            type: Function,
            default: val => val,
            required: false
        },
        customData:{
            type:Array,
            default: () => []
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },
    mixins: [Select],

    methods: {
        loadData(param) {
            let codedicApi=new BufferedCodedicApi(param.module);
            codedicApi.reqData(param, data => {
                let result= this.customData.concat(data);
                this.$emit('loadEnd',result);
                this.dataSource = this.onLoad(result);
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
