// ui-jz-v4-common/src/components/qb-base/table-drop.js
/**
 * 下拉表格
 * @Author:
 * @Date:   2019-04-02
 * @Project qb
 */

import { TableDeop } from "ui-component-v4";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api
export default {
    name: 'QbTableDeop',

    props: {
        data: {
            type: Array,
            default: () => [],
            required: false
        },
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // sql
        sql: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            required: false
        },
        customData:{
            type:Array,
            default: () => []
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },
    mixins: [TableDeop],

    methods: {
        loadData(params) {
            let codedicApi=new BufferedCodedicApi(params.module);
            codedicApi.reqData(params, data => {
                let result= this.customData.concat(data);
                this.$emit('loadEnd',result);
                this.dataSource = result;
            });
        },

        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
