// ui-jz-v4-common/src/components/qb-base/transfer.js
/**
 * 左右穿梭框
 * @Author:
 * @Date:   2019-03-06
 * @Project qb
 */
import { Transfer } from "ui-component-v4";
import { BufferedCodedicApi } from '../../api/qb-base/BufferedCodedicApi'//表码翻译api
// export default Transfer;
export default {
    name: 'QbTransfer',

    props: {
        url: {
            type: String,
            default: "",
            required: false
        },
        value: {
            type: Array,
            default() {
                return [];
            },
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        //不使用缓存
        noCache: {
            type: Boolean,
            default: true,
            require: false
        },
        loadingShow:{
            type: Boolean,
            default: false
        }
    },

    mixins: [Transfer],

    methods: {
        loadData(params) {
            let codedicApi=new BufferedCodedicApi(params.module);
            codedicApi.reqData(params, data => {
                this.$emit('loadEnd',data);
                this.dataSource = data;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
