// @qb/qs-icccd/src/pages/scenes/add/data-save.cjs
/**
 * 场景模块 - 组件配置和模拟数据
 * Created by hecg on 2020/10/13.
 */
// 配置 场景ID
const sceneId = 'nzdr_gd';
// 配置 模块
const MODULE = '';
// 配置 表码
const bm = {
  BKDX: 'BM_BKDX_HMLX',
  BKLX: 'BM_BKSQ_BKLB',
  BKFW: 'BM_BKSQ_BKFW',
  BKJB: 'BM_BKSQ_BKJB',
  BKZQ: 'BM_BKSQ_BKQX',
  CZYQ: 'BM_BKSQ_CZCS',
  YJQSR: 'BM_YJFW_YJQSR', // TODO 暂时不做
  YJFS: 'BM_BKSQ_YJFS'
};

const util = {
  sort(i1, i2) {
    return i1.order - i2.order;
  },
  generateKey(len = 32) {
    const $chars = 'abcdefhijkmnprstwxyz2345678';
    const result = [];
    for (let i = 0; i < len; i++) {
      result.push($chars.charAt((Math.random() * 27) << 0));
    }
    return result.join('');
  }
};
const moduleKey = '-' + util.generateKey(10);

// 场景模块列表
const T_MANAGE_SCENE_MODULE = [
  // 模块配置
  { name: '布控对象参数', id: 'bkdx' + moduleKey, code: '' },
  {
    name: '申请信息参数',
    id: 'sqxx' + moduleKey,
    code:
      'const unWatch = this.$watch(() => {return this.componentModel.bkjb;},bkjb => {if (bkjb) {this.$set(this.componentModel, "czcsyq", bkjb);if (bkjb == "1") {this.list.forEach(item => {if (item.formBind.prop == "zfyjxx") {item.formBind.rules.push({ required: true, message: "执法依据不能为空" });}});} else {this.list.forEach(item => {if (item.formBind.prop == "zfyjxx") {item.formBind.rules.splice(0,1);}});}}},{ deep: true, immediate: true });return () => {unWatch();};'
  },
  { name: '预警方式参数', id: 'yjfs' + moduleKey, code: '' }
];
// 场景模块-左侧组件列表
const T_MANAGE_MODULE_COMPONENT = [
  { modelId: 'bkdx' + moduleKey, componentId: 'bkdx', label: '布控对象', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'rwmc', label: '任务名称', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'bkfw', label: '布控范围', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'bkjb', label: '布控级别', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'czyq', label: '处置要求', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'bkzq', label: '布控周期', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'bklx', label: '布控类别', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'bksy', label: '布控事由', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'zfyj', label: '执法依据', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'yjtj', label: '预警条件', type: '1', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'wb', label: '文本', type: '2', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'wj', label: '文件', type: '2', order: 1 },
  { modelId: 'sqxx' + moduleKey, componentId: 'tp', label: '图片', type: '2', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'yjfs', label: '预警方式', type: '1', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'yjjsr', label: '预警接收人', type: '1', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'yjsj', label: '预警手机', type: '1', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'wb', label: '文本', type: '2', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'wj', label: '文件', type: '2', order: 1 },
  { modelId: 'yjfs' + moduleKey, componentId: 'tp', label: '图片', type: '2', order: 1 }
];
// 组件属性配置表
const T_MANAGE_COMPONENT = [
  {
    id: 'bkdx',
    option: {
      // 禁止删除
      noDelete: true,
      // 组件选项绑定的 key
      watchKey: 'selectData',
      // 组件配置项
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控对象',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'selectPlaceholder',
          label: '提示文字（布控对象类型）',
          is: 'el-input',
          value: '请选择布控对象类型',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'inputPlaceholder',
          label: '提示文字（布控对象信息）',
          is: 'el-input',
          value: '请输入布控对象信息',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'selectData',
          label: '选项',
          is: 'dg-select',
          value: ['111', 'RLZP', 'SJ', 'CPH', 'WX', 'QQ'],
          bind: {
            clearable: true,
            multiple: true,
            module: MODULE,
            code: bm.BKDX
          }
        },
        {
          prop: 'value',
          label: '默认值',
          is: 'dg-select',
          value: '111',
          bind: {
            module: MODULE,
            clearable: true
          }
        }
      ],
      formBind: {
        // 实际绑定到实体上的属性字段
        prop: 'bkdx',
        rules: [{ required: true, message: '布控对象不能为空！' }]
      },
      bind: {
        is: 'scenes-bkdx',
        value: '01'
      }
    }
  },
  {
    id: 'rwmc',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '申请单名称',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请输入申请单名称'
        }
      ],
      formBind: {
        prop: 'sqmc',
        rules: [{ required: true, message: '申请单名称不能为空！' }]
      },
      bind: {
        is: 'el-input',
        value: ''
      }
    }
  },
  {
    id: 'bkfw',
    option: {
      // watchKey: 'data',
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控范围',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        // {
        //   prop: 'data',
        //   label: '选项',
        //   is: 'dg-select',
        //   value: '',
        //   bind: {
        //     clearable: true,
        //     multiple: true,
        //     module: MODULE,
        //     code: bm.BKFW
        //   }
        // },
        {
          use: 'rules',
          label: '是否必填',
          prop: 'required',
          is: 'el-switch',
          value: true,
          message: '布控范围不能为空！',
          bind: { disabled: true }
        }
      ],
      formBind: {
        prop: 'bkfwzh',
        rules: [{ required: true, message: '布控范围不能为空！' }]
      },
      bind: {
        is: 'nationwide-selection',
        module: MODULE,
        code: bm.BKFW,
        value: ''
      }
    }
  },
  {
    id: 'bkjb',
    option: {
      watchKey: 'data',
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控级别',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请选择布控级别'
        },
        {
          prop: 'data',
          label: '选项',
          is: 'dg-select',
          value: '1,2,3',
          bind: {
            clearable: true,
            multiple: true,
            module: MODULE,
            code: bm.BKJB
          }
        },
        {
          prop: 'value',
          label: '默认值',
          is: 'dg-select',
          value: '3',
          bind: { clearable: true }
        }
      ],
      formBind: {
        prop: 'bkjb',
        rules: [{ required: true, message: '布控级别不能为空！' }]
      },
      bind: {
        is: 'dg-radio-group',
        value: ''
      }
    }
  },
  {
    id: 'czyq',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '处置要求',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        }
      ],
      formBind: {
        prop: 'czcsyq',
        label: '处置要求'
      },
      bind: {
        // is: 'qs-translate',
        is: 'qs-input',
        module: MODULE,
        code: bm.CZYQ,
        value: '1'
      }
    }
  },
  {
    id: 'bklx',
    option: {
      // watchKey: 'data',
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控类别',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请选择布控类别'
        }
        // {
        //   prop: 'data',
        //   label: '选项',
        //   is: 'dg-select',
        //   value: '11,12,13',
        //   bind: {
        //     clearable: true,
        //     multiple: true,
        //     module: MODULE,
        //     code: bm.BKLX
        //   }
        // },
        // {
        //   prop: 'value',
        //   label: '默认值',
        //   is: 'dg-select',
        //   value: '',
        //   bind: { clearable: true }
        // }
      ],
      formBind: {
        prop: 'bklb',
        rules: [{ required: true, message: '布控类别不能为空！' }]
      },
      bind: {
        // is: 'dg-select',
        is: 'dg-tree-drop',
        module: MODULE,
        filterable: true,
        clearable: true,
        'lazy-load': true,
        'tree-props': {
          // 展开联动、收缩不联动
          isToggleContact: true,
          isHighLight: true,
          nodeKey: 'value',
          checkOnClickNode: true
        },
        'value-name': 'value',
        'label-name': 'label',
        'check-strictly': false,
        props: {
          value: 'value'
        },
        code: bm.BKLX,
        value: ''
      }
    }
  },
  {
    id: 'bkzq',
    option: {
      watchKey: 'data',
      defaultOptionKey: 'activeValue',
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控周期',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'data',
          label: '选项',
          is: 'dg-select',
          value: '30,60,90',
          bind: {
            clearable: true,
            multiple: true,
            module: MODULE,
            code: bm.BKZQ
          }
        },
        {
          prop: 'activeValue',
          label: '默认值',
          is: 'dg-select',
          value: '',
          bind: {
            data: [],
            clearable: true
          }
        }
      ],
      formBind: {
        prop: 'bkqx',
        rules: [{ required: true, message: '布控周期不能为空！' }]
      },
      bind: {
        is: 'qs-cycle-pick',
        code: bm.BKZQ,
        activeValue: '',
        data: [],
        value: ''
      }
    }
  },
  {
    id: 'bksy',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '布控事由',
          use: 'form-item',
          bind: { placeholder: '请输入布控事由' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请输入布控事由',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'maxlength',
          label: '最大长度',
          is: 'el-input-number',
          value: 500,
          bind: { placeholder: '请输入最大长度' }
        },
        {
          prop: 'show-word-limit',
          label: '显示输入字数统计',
          is: 'el-switch',
          value: true,
          bind: {}
        }
      ],
      formBind: {
        prop: 'bkyy',
        rules: [{ required: true, message: '布控事由不能为空！' }]
      },
      bind: {
        is: 'el-input',
        type: 'textarea',
        row: 3,
        value: ''
      }
    }
  },
  {
    id: 'yjtj',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '预警条件',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          use: 'rules',
          label: '是否必填',
          prop: 'required',
          is: 'el-switch',
          value: true,
          message: '预警条件不能为空！'
        }
      ],
      formBind: {
        prop: 'yjtj',
        rules: []
      },
      bind: {
        is: 'warning-rule',
        module: MODULE,
        value: {}
      }
    }
  },
  {
    id: 'zfyj',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '执法依据',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'tip-msg',
          label: '提示文字',
          is: 'el-input',
          value:
            '请按规范要求上传立案决定书、临控申请表、移交承诺书等执法依据，上传文件个数限制为 10 个，单个文件大小为 2M',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'limit',
          label: '文件限制个数',
          is: 'dg-input-number',
          value: 10,
          bind: { placeholder: '请输入文件限制个数', step: 1 }
        },
        {
          prop: 'size',
          label: '文件限制大小',
          is: 'el-input',
          value: '2M',
          bind: { placeholder: '请输入文件限制大小', step: 1 }
        }
      ],
      formBind: {
        prop: 'zfyjxx',
        label: '执法依据',
        rules: []
      },
      bind: {
        is: 'dg-upload',
        'list-type': 'button',
        action: 'http://10.201.3.21:8891/qz-bksq/api/file/upload',
        'tip-msg': '',
        multiple: true,
        limit: 10,
        size: '2M',
        value: []
      }
    }
  },
  {
    id: 'yjfs',
    option: {
      watchKey: 'data',
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '预警方式',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'data',
          label: '选项',
          is: 'dg-select',
          value: '01,02,03',
          bind: {
            clearable: true,
            multiple: true,
            module: MODULE,
            code: bm.YJFS
          }
        },
        {
          prop: 'value',
          label: '默认值',
          is: 'dg-select',
          value: '',
          bind: { clearable: true, multiple: true }
        }
      ],
      formBind: {
        prop: 'yjfs',
        rules: [{ required: true, message: '预警方式不能为空！' }]
      },
      bind: {
        is: 'dg-select',
        clearable: true,
        multiple: true,
        module: MODULE,
        value: ''
      }
    }
  },
  {
    id: 'yjjsr',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '预警接收人',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'filterPlaceholder',
          label: '搜索框提示文字',
          is: 'el-input',
          value: '请输入姓名',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'code',
          label: '选项',
          is: 'dg-select',
          value: '',
          bind: {
            module: MODULE,
            code: bm.YJQSR,
            placeholder: '请选择人员表码'
          }
        }
        // {
        //   prop: 'data',
        //   label: '选项',
        //   is: 'qs-tree-drop',
        //   value: '',
        //   bind: {
        //     module: MODULE,
        //     code: bm.YJQSR,
        //     props: { value: 'id', label: 'name', children: 'children', isLeaf: 'leaf' },
        //     'tree-props': {
        //       // 展开联动、收缩不联动
        //       isToggleContact: true,
        //       // 高亮检索
        //       isHighLight: true,
        //       nodeKey: 'id',
        //       lazy: true,
        //       // load: this.loadNode,
        //       showCheckbox: true
        //     },
        //     placeholder: '请选择人员',
        //     clearable: true,
        //     filterable: true,
        //     multiple: true
        //   }
        // }
      ],
      formBind: {
        prop: 'yjjsr'
      },
      bind: {
        is: 'yjjsr-transfer',
        leftTitle: '全部接收人',
        rightTitle: '已选接收人',
        url: '/dict/translation-to-tree',
        module: MODULE,
        value: ''
      }
    }
  },
  {
    id: 'yjsj',
    option: {
      configs: [
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '预警手机',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请输入其他预警手机号码，多个用“,”隔开',
          bind: { placeholder: '请输入提示文字' }
        }
      ],
      formBind: {
        prop: 'yjsj'
      },
      bind: {
        is: 'el-input',
        type: 'textarea',
        row: 3,
        value: ''
      }
    }
  },
  {
    id: 'wb',
    option: {
      configs: [
        {
          prop: 'prop',
          label: '绑定字段属性（必填）',
          is: 'el-input',
          value: 'wb',
          use: 'form-item',
          bind: { placeholder: '请输入绑定字段属性' }
        },
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '文本',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'placeholder',
          label: '提示文字',
          is: 'el-input',
          value: '请输入文本',
          bind: { placeholder: '请输入提示文字' }
        },
        {
          prop: 'rows',
          label: '行数',
          is: 'el-input-number',
          value: 1,
          bind: { placeholder: '请输入行数' }
        },
        {
          prop: 'maxlength',
          label: '最大长度',
          is: 'el-input-number',
          value: 500,
          bind: { placeholder: '请输入最大长度' }
        },
        {
          prop: 'show-word-limit',
          label: '显示输入字数统计',
          is: 'el-switch',
          value: true,
          bind: {}
        },
        {
          prop: 'required',
          label: '是否必填',
          is: 'el-switch',
          value: false,
          use: 'rules',
          message: '文本不能为空！',
          bind: {}
        }
      ],
      formBind: {},
      bind: {
        is: 'el-input',
        type: 'textarea',
        value: ''
      }
    }
  },
  {
    id: 'wj',
    option: {
      configs: [
        {
          prop: 'prop',
          label: '绑定字段属性（必填）',
          is: 'el-input',
          value: 'wj',
          use: 'form-item',
          bind: { placeholder: '请输入绑定字段属性' }
        },
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '文件',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'required',
          label: '是否必填',
          is: 'el-switch',
          value: false,
          use: 'rules',
          message: '请选择文件！',
          bind: {}
        }
      ],
      formBind: {},
      bind: {
        is: 'dg-upload',
        multiple: true,
        'list-type': 'button',
        'tip-msg': '上传文件个数限制为 5 个，大小为 10M',
        limit: 5,
        size: '10m',
        action: '/文件上传地址',
        value: ''
      }
    }
  },
  {
    id: 'tp',
    option: {
      configs: [
        {
          prop: 'prop',
          label: '绑定字段属性（必填）',
          is: 'el-input',
          value: 'tp',
          use: 'form-item',
          bind: { placeholder: '请输入绑定字段属性' }
        },
        {
          prop: 'label',
          label: '标题',
          is: 'el-input',
          value: '图片',
          use: 'form-item',
          bind: { placeholder: '请输入标题' }
        },
        {
          prop: 'required',
          label: '是否必填',
          is: 'el-switch',
          value: false,
          use: 'rules',
          message: '请选择图片！',
          bind: {}
        }
      ],
      formBind: {},
      bind: {
        is: 'dg-upload',
        'list-type': 'avatar',
        action: '/照片上传地址',
        value: ''
      }
    }
  }
];
// 场景模块-组件配置表 页面中间部分
const T_MANAGE_MODULE_CONFIG = [
  {
    modelId: 'bkdx' + moduleKey,
    componentId: 'bkdx',
    order: 0,
    option: {
      formBind: { prop: 'bkdx', rules: [{ required: true, message: '布控对象不能为空！' }], label: '布控对象' },
      bind: {
        is: 'scenes-bkdx',
        value: '111',
        selectPlaceholder: '请选择布控对象类型',
        inputPlaceholder: '请输入布控对象信息',
        selectData: [
          { value: '111', label: '身份证号' },
          { value: 'RLZP', label: '照片' },
          { value: 'CPH', label: '车牌号' },
          { value: 'SJ', label: '手机号' },
          { value: 'WX', label: '微信号' },
          { value: 'QQ', label: 'QQ号' }
        ]
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'rwmc',
    order: 0,
    option: {
      formBind: { prop: 'sqmc', rules: [{ required: true, message: '申请单名称不能为空！' }], label: '申请单名称' },
      bind: { is: 'el-input', value: '', placeholder: '请输入申请单名称' }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'bkfw',
    order: 1,
    option: {
      formBind: { prop: 'bkfwzh', rules: [{ required: true, message: '布控范围不能为空！' }], label: '布控范围' },
      bind: { is: 'nationwide-selection', module: '', code: 'BM_BKSQ_BKFW', value: '' }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'bklx',
    order: 2,
    option: {
      formBind: { prop: 'bklb', rules: [{ required: true, message: '布控类别不能为空！' }], label: '布控类别' },
      bind: {
        is: 'dg-tree-drop',
        module: '',
        filterable: true,
        clearable: true,
        'lazy-load': true,
        'tree-props': {
          isHighLight: true,
          nodeKey: 'value'
        },
        'value-name': 'value',
        'label-name': 'label',
        'check-strictly': false,
        props: { value: 'value' },
        code: 'BM_BKSQ_BKLB',
        value: '',
        placeholder: '请选择布控类别'
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'bkjb',
    order: 3,
    option: {
      formBind: { prop: 'bkjb', rules: [{ required: true, message: '布控级别不能为空！' }], label: '布控级别' },
      bind: {
        is: 'dg-radio-group',
        value: '3',
        placeholder: '请选择布控级别',
        data: [
          { value: '1', label: '一级（抓捕）' },
          { value: '2', label: '二级（关注）' },
          { value: '3', label: '三级（掌握轨迹）' }
        ]
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'czyq',
    order: 4,
    option: {
      formBind: { prop: 'czcsyq', label: '处置要求' },
      bind: { is: 'qs-input', module: '', code: 'BM_BKSQ_CZCS', value: '1' }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'bkzq',
    order: 5,
    option: {
      formBind: { prop: 'bkqx', rules: [{ required: true, message: '布控周期不能为空！' }], label: '布控周期' },
      bind: {
        is: 'qs-cycle-pick',
        code: 'BM_BKSQ_BKQX',
        activeValue: '30',
        data: [
          { value: '30', label: '30天' },
          { value: '60', label: '60天' },
          { value: '90', label: '90天' }
        ],
        value: ''
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'bksy',
    order: 6,
    option: {
      formBind: { prop: 'bkyy', rules: [{ required: true, message: '布控事由不能为空！' }], label: '布控事由' },
      bind: {
        is: 'el-input',
        type: 'textarea',
        row: 3,
        value: '',
        placeholder: '请输入布控事由',
        maxlength: 500,
        'show-word-limit': true
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'zfyj',
    order: 7,
    option: {
      formBind: { prop: 'zfyjxx', label: '执法依据', rules: [] },
      bind: {
        is: 'dg-upload',
        'list-type': 'button',
        action: 'http://10.201.3.21:8891/qz-bksq/api/file/upload',
        'tip-msg':
          '请按规范要求上传立案决定书、临控申请表、移交承诺书等执法依据，上传文件个数限制为 10 个，单个文件大小为 2M',
        multiple: true,
        limit: 10,
        size: '2M',
        value: []
      }
    }
  },
  {
    modelId: 'sqxx' + moduleKey,
    componentId: 'yjtj',
    order: 8,
    option: {
      formBind: { prop: 'yjtj', rules: [], label: '预警条件' },
      bind: { is: 'warning-rule', module: '', value: {} }
    }
  },
  {
    modelId: 'yjfs' + moduleKey,
    componentId: 'yjfs',
    order: 0,
    option: {
      formBind: { prop: 'yjfs', rules: [{ required: true, message: '预警方式不能为空！' }], label: '预警方式' },
      bind: {
        is: 'dg-select',
        clearable: true,
        multiple: true,
        module: '',
        value: ['40'],
        data: [
          { value: '40', label: '系统消息提示' },
          { value: '10', label: '手机短信提示' },
          { value: '60', label: '邮箱发送提示' }
        ]
      }
    }
  },
  {
    modelId: 'yjfs' + moduleKey,
    componentId: 'yjsj',
    order: 1,
    option: {
      formBind: { prop: 'yjsj', label: '预警手机' },
      bind: {
        is: 'el-input',
        type: 'textarea',
        row: 3,
        value: '',
        placeholder: '请输入其他预警手机号码，多个用“,”隔开'
      }
    }
  }
];
// 发布之后的数据
// const T_MANAGE_SCENE_RELEASE = {
//   id: util.generateKey(),
//   sceneId,
//   options: []
// };

// 获取场景模块
function getSceneModelList() {
  return T_MANAGE_SCENE_MODULE;
}
// 左侧列表
function getLeftList(modelId) {
  const list = T_MANAGE_MODULE_COMPONENT.filter(e => e.modelId === modelId);
  list.sort(util.sort);
  list.forEach(e => {
    e.id = util.generateKey();
  });
  return list;
}
// 右侧组件属性配置项
function getComponentOption(componentId) {
  const item = T_MANAGE_COMPONENT.find(e => e.id === componentId);
  return item && item.option;
}
// 中间预览
function getPreview(modelId) {
  const list = T_MANAGE_MODULE_CONFIG.filter(e => e.modelId === modelId);
  list.sort(util.sort);
  return list.map(e => {
    const config = getComponentOption(e.componentId);
    const item = {
      id: util.generateKey(),
      ...e,
      ...config
    };
    if (item.bind) {
      item.bind = { ...item.bind, ...item.option.bind };
    }
    if (item.formBind) {
      item.formBind = { ...item.formBind, ...item.option.formBind };
    }
    delete item.option;
    return item;
  });
}
// 发布之后的数据
function getPublish() {
  return T_MANAGE_SCENE_MODULE.map(item => {
    const options = getPreview(item.id).map(({ order, bind, formBind }) => {
      return { order, bind, formBind };
    });
    return { ...item, options };
  });
}

module.exports = {
  util,
  sceneId,
  getSceneModelList,
  getLeftList,
  getComponentOption,
  getPreview,
  getPublish,
  T_MANAGE_SCENE_MODULE,
  T_MANAGE_MODULE_COMPONENT,
  T_MANAGE_COMPONENT,
  T_MANAGE_MODULE_CONFIG
};
