// ui-component-v4/packages/theme-chalk/gulpfile.js
// update
'use strict';

const { series, src, dest } = require('gulp');
// const sass = require('gulp-sass');
const autoprefixer = require('gulp-autoprefixer');
const cssmin = require('gulp-cssmin');

var sass = require('gulp-sass');
sass.compiler = require('node-sass');

function compile() {
    return src('./src/*.scss')
        .pipe(sass.sync())
        .pipe(
            autoprefixer({
                // +lutz- 版本问题 [code] browsers: ['ie > 9', 'last 2 versions']
                overrideBrowserslist: ['ie > 9', 'last 2 versions'], // lutz+
                cascade: false
            })
        )
        .pipe(cssmin())
        .pipe(dest('./lib'));
}

function compileIndex() {
    return src('./src/index.scss')
        .pipe(sass())
        .pipe(
            autoprefixer({
                // +lutz- 版本问题 [code] browsers: ['ie > 9', 'last 2 versions']
                overrideBrowserslist: ['ie > 9', 'last 2 versions'], // lutz+
                cascade: false
            })
        )
        .pipe(cssmin())
        .pipe(dest('./lib'));
}

function copyfont() {
    // +lutz- 引用 element-ui 的 font
    // [code] src('./src/fonts/**').pipe(cssmin()).pipe(dest('./lib/fonts'));
    return src('./src/element-scss/fonts/**')
        .pipe(cssmin())
        .pipe(dest('./lib/fonts'));
}

// lutz+ 拷贝 icon
function copyicon() {
    return src('./src/icon/**')
        .pipe(cssmin())
        .pipe(dest('./lib/icon'));
}

// lutz+ 拷贝 images
function copyimage() {
    return src('./src/images/**')
    .pipe(cssmin())
    .pipe(dest('./lib/images'));
}

exports.build = series(compile, copyfont, copyicon, copyimage);
exports.buildIndex = series(compileIndex);
