// ui-component-v4/packages/form-item/src/message-CN.js
/*
 * @Author: lutz 
 * @Date: 2019-04-10 13:21:57 
 * @Last Modified by:   lutz 
 * @Last Modified time: 2019-06-10 13:21:57 
 */

export default {
  default: '字段上验证处在错误%s',
  required: '%s是必须的',
  enum: '%s必须是其中之一%s',
  whitespace: '%s不能为空',
  date: {
    format: '%s日期%s是无效格式%s',
    parse: '%s日期格式不能被解析,%s是无效的 ',
    invalid: '%s无效日期格式：%s',
  },
  types: {
    string: '%s不是一个字符(%s)类型',
    method: '%s不是一个%s(function)类型',
    array: '%s不是一个数组(%s)类型',
    object: '%s不是一个对象(%s)类型',
    number: '%s不是一个数字(%s)类型',
    date: '%s不是一个日期(%s)类型',
    boolean: '%s不是一个布尔(%s)类型',
    integer: '%s不是一个整型(%s)类型',
    float: '%s不是一个浮点(%s)类型',
    regexp: '%s不是一个有效正则(%s)类型',
    email: '%s不是一个有效值%s格式',
    url: '%s不是一个有效值%s格式',
    hex: '%s不是一个有效值%s格式',
  },
  string: {
    len: '%s必须等于%s字节',
    min: '%s必须大于%s字符',
    max: '%s必须小于%s字节',
    range: '%s必须在%s到%s字符内',
  },
  number: {
    len: '%s必须等于%s',
    min: '%s必须大于%s',
    max: '%s必须小于%s',
    range: '%s必须在%s到%s范围内',
  },
  array: {
    len: '%s数组长度必须等于%s',
    min: '%s数组长度必须大于%s',
    max: '%s数组长度必须小于%s',
    range: '%s数组长度必须在%s到%s范围内',
  },
  pattern: {
    mismatch: '%s的值%s不能正确匹配%s',
  }
}
