// @qb/qs-icccd/src/store/modules/backgroundState.js
const bgState = {
  state: {
    selfDomActive: 0,
    simpleActive: 0,
    // bgActive:0,
    themeActive: 0
  },
  mutations: {
    SELFDOMACTIVE: (state, index) => {
      state.selfDomActive = index;
    },
    SIMPLEACTIVE: (state, index) => {
      state.simpleActive = index;
    },
    // BGACTIVE: (state, index) => {
    //     //这边执行提交axios或者直接执行
    //     state.bgActive = index
    // },
    THEMEACTIVE: (state, index) => {
      // 这边执行提交axios或者直接执行
      state.themeActive = index;
    }
  },
  actions: {
    SubSelfDomActive({ commit }, index) {
      commit('SELFDOMACTIVE', index);
    },
    SubSimpleActive({ commit }, index) {
      commit('SIMPLEACTIVE', index);
    },
    // SubBgActive({ commit }, index) {
    //     commit("BGACTIVE", index);
    // },
    SubThemeActive({ commit }, index) {
      commit('THEMEACTIVE', index);
    }
  },
  getters: {
    getSelfDomActive: state => state.selfDomActive,
    getSimpleActive: state => state.simpleActive,
    // getBgActive: state => state.bgActive,
    getThemeActive: state => state.themeActive
  }
};

export default bgState;
