// @qb/qs-icccd/src/pages/scenes/request.js
/**
 * Created by hecg on 2020/10/27.
 */
import request from '@icccdPath/utils/request';

export function sceneModel(id) {
  return request({
    url: `/v1/scene/model/${id}`
  });
}

/**
 * 查询模块下的已选组件
 * @param modelId {String} 模块ID
 */
export function loadConfig(modelId) {
  return request({
    url: `/v1/scene/model/loadConfig/${modelId}`
  });
}

/**
 * 查询模块下的可选组件
 * @param modelId {String} 模块ID
 */
export function modelComponentList(modelId) {
  return request({
    url: `/v1/scene/model/component-list/${modelId}`
  });
}

/**
 * 查询组件配置项
 * @param componentId {String} 组件ID
 */
export function modelComponent(componentId) {
  return request({
    url: `/v1/scene/model/component/${componentId}`
  });
}

/**
 * 场景模块已选组件保存
 * @param data {Array} 模块数据
 */
export function modelComponentListSave(data) {
  return request({
    url: '/v1/scene/model-component-list/save',
    method: 'post',
    data
  });
}

/**
 * 场景发布
 * @param id {String} 场景ID
 */
export function releasePublish(id) {
  return request({
    url: `/v1/scene/release/publish/${id}`,
    method: 'post'
  });
}
