// ui-jz-v4-common/src/components/jz-base/cascader.js
/**
 * 级联选择框
 * @Author: huangjq
 * @Date:   2019-03-12
 * @Project jz
 */
import { Cascader } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: Cascader.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        options: {
            type: Array,
            default: () => [],
            require: false
        }
    },

    watch: {
        // 表码
        code() {
            this.loadData(this.$props);
        },
        // 枚举
        enum() {
            this.loadData(this.$props);
        },
        // url
        url(){
            this.loadData(this.$props);
        }
    },

    mixins: [Cascader],

    methods: {
        getLabel() {
            return this.dataSource
                .filter(item => this.value.split(",").includes(item[this.valueName]))
                .map(item => item[this.labelName])
                .join(",");
        },
        // 加载数据
        loadData(params) {
            DataOp.fetchData(params, data => {
                this.dataSource = data;
            });
        },
        // 初始化数据
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
