// ui-jz-v4-common/src/components/jz-base/checkbox-group.js
/**
 * 多选框组
 * @Author: huangjq
 * @Date:   2019-02-27
 * @Project jz
 */
import { CheckboxGroup } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: CheckboxGroup.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        customData:{
            type:Array,
            default: () => []
        }
    },

    watch: {
        // 表码
        code() {
            this.loadData(this.$props);
        },
        // 枚举
        enum() {
            this.loadData(this.$props);
        },
        // url
        url(){
            this.loadData(this.$props);
        }
    },

    mixins: [CheckboxGroup],

    methods: {
        getLabel() {
            return this.dataSource
                .filter(item => this.value.split(",").includes(item[this.valueName]))
                .map(item => item[this.labelName])
                .join(",");
        },
        loadData(params) {
            DataOp.fetchData(params, data => {
                let result= this.customData.concat(data);
                this.dataSource = result;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
