// ui-jz-v4-common/src/components/jz-base/form-column.js
/**
 * 表单表格列
 * @Author: huangjq
 * @Date:   2019-06-24
 * @Project jz
 */

import { FormColumn } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: FormColumn.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        sortable: {
            type: [String, Boolean],
            default: false,
            required: false
        },
        dataType: String,
        format: String,
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        editPropsData: {
            type: Array,
            default: () => [],
            require: false
        }
    },

    watch: {
        // 表码
        code() {
            this.loadCode(this.$props);
        },
        // 枚举
        enum() {
            this.loadCode(this.$props);
        },
        // url
        url(){
            this.loadCode(this.$props);
        }
    },

    computed: {},

    mixins: [FormColumn],

    mounted() {},

    methods: {
        _getCmpPropsCallback(attrs) {
            if (attrs.type === "opt") {
                attrs.sortable = false;
            }
        },

        _getCmpProps(bindProps) {
            this.url && (bindProps.url = this.url);
            this.code && (bindProps.code = this.code);
            this.enum && (bindProps.enum = this.enum);
            this.module && (bindProps.module = this.module);
        },

        codeFormatter({ column, row }) {
            if (this.dataType === "date") {
                return this.$moment(row[column.property]).format(this.format);
            } else {
                return row[column.property];
            }
        },
        /**
         * 加载表码
         */
        loadCode() {
            if (this.code || this.enum || this.url) {
                DataOp.fetchData(
                    this.$props,
                    data => {
                        this.codeObj = data;
                    },
                    "codeObj"
                );
            }
        }
    },

    created() {
        this.loadCode();
    }
};
