// ui-jz-v4-common/src/components/jz-base/line-column.js
/**
 * 此处填写文件描述
 * @Author: huangjq
 * @Date:   2019-03-07
 * @Project jz
 */

import { LineColumn } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: LineColumn.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        sortable: {
            type: [String, Boolean],
            default: false,
            required: false
        },
        // :fold="false" show-overflow-tooltip
        showOverflowTooltip: {
            type: Boolean,
            default: true,
            required: false
        },
        fold: {
            type: Boolean,
            default: false,
            required: false
        },
        dataType: String,
        format: String,
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        editPropsData: {
            type: Array,
            default: () => [],
            require: false
        }
    },

    watch: {
        // 表码
        code() {
            this.loadCode(this.$props);
        },
        // 枚举
        enum() {
            this.loadCode(this.$props);
        },
        // url
        url(){
            this.loadCode(this.$props);
        }
    },

    computed: {},

    mixins: [LineColumn],

    mounted() {},

    methods: {
        _getCmpPropsCallback(attrs) {
            if (attrs.type === "operation-tpl") {
                attrs.sortable = false;
            }
        },

        _getCmpProps(bindProps) {
            Object.assign(
                bindProps,
                { code: this.code },
                { enum: this.enum },
                { url: this.url },
                { module: this.module }
            );
        },

        /**
         * 列数据格式化
         * @param column
         * @param row
         * @return {*}
         */
        codeFormatter({ column, row }) {
            if (this.dataType === "date") {
                return this.$moment(row[column.property]).format(this.format);
            } else {
                return row[column.property];
            }
        },

        /**
         * 加载表码
         */
        loadCode() {
            if (this.code || this.enum || this.url) {
                DataOp.fetchData(
                    this.$props,
                    data => {
                        this.codeObj = data;
                    },
                    "codeObj"
                );
            }
        }
    },

    created() {
        this.loadCode();
    }
};
