// ui-jz-v4-common/src/components/jz-base/radio-group.js
/**
 * 单选框组
 * @Author: huangjq
 * @Date:   2019-02-25
 * @Project jz
 */
import { RadioGroup } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: RadioGroup.name,

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        },
        customData:{
            type:Array,
            default: () => []
        }
    },

    watch: {
        // 表码
        code() {
            this.loadData(this.$props);
        },
        // 枚举
        enum() {
            this.loadData(this.$props);
        },
        // url
        url(){
            this.loadData(this.$props);
        }
    },

    mixins: [RadioGroup],

    methods: {
        getLabel() {
            let obj = this.dataSource.find(item => item[this.valueName] === this.value);
            return obj ? obj[this.labelName] : "";
        },
        loadData(params) {
            DataOp.fetchData(params, data => {
                let result= this.customData.concat(data);
                this.dataSource = result;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
