// ui-jz-v4-common/src/components/jz-base/table-column.js
/**
 * 此处填写文件描述
 * @Author: huangjq
 * @Date:   2019-03-07
 * @Project jz
 */

import { TableColumn } from "ui-component-v4";
import DataOp from "../../utils/zt-data";

export default {
    name: "DgTableColumn",

    props: {
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        sortable: {
            type: [String, Boolean],
            default: false,
            required: false
        },
        // :fold="false" show-overflow-tooltip
        showOverflowTooltip: {
            type: Boolean,
            default: false,
            required: false
        },
        fold: {
            type: Boolean,
            default: false,
            required: false
        },
        dataType: String,
        format: String,
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            require: false
        }
    },

    watch: {
        // 表码
        code() {
            this.loadCode(this.$props);
        },
        // 枚举
        enum() {
            this.loadCode(this.$props);
        },
        // url
        url(){
            this.loadCode(this.$props);
        }
    },

    computed: {},

    mixins: [TableColumn],

    mounted() {
    },

    methods: {
        _getCmpPropsCallback(attrs) {
            if (attrs.type === "operation-tpl") {
                attrs.sortable = false;
            }
        },

        _getCmpProps(bindProps) {
            bindProps.scope.column.sortable = false;
        },

        /**
         * 格式化列
         * @param value
         * @param column
         * @param row
         * @return {*}
         * @private
         */
        _formatValue(value) {
            if (this.dataType === "date" && value) {
                if (this.$moment(value).isValid()) {
                    value = this.$moment(value).format(this.format);
                } else {
                    let tmp = value;
                    let length = value.length;
                    tmp = tmp.split("");
                    if (length > 4) {
                        tmp.splice(4, 0, "-");
                    }
                    if (length > 6) {
                        tmp.splice(7, 0, "-");
                    }
                    if (length > 8) {
                        tmp.splice(10, 0, " ");
                    }
                    if (length > 10) {
                        tmp.splice(13, 0, ":");
                    }
                    if (length > 12) {
                        tmp.splice(16, 0, ":");
                    }
                    tmp = tmp.join("");
                    if (this.$moment(tmp).isValid()) {
                        value = this.$moment(tmp).format(this.format);
                    }
                }
            }
            return value;
        },

        /**
         * 加载表码
         */
        loadCode() {
            if (this.code || this.enum || this.url) {
                DataOp.fetchData(
                    this.$props,
                    data => {
                        this.codeObj = data;
                    },
                    "codeObj"
                );
            }
        }
    },

    created() {
        this.loadCode();
    }
};
