// ui-jz-v4-common/src/components/jz-base/transfer.js
/**
 * 左右穿梭框
 * @Author: huangjq
 * @Date:   2019-03-06
 * @Project jz
 */
import { Transfer } from "ui-component-v4";
import DataOp from "../../utils/zt-data";
// export default Transfer;
export default {
    name: Transfer.name,

    props: {
        url: {
            type: String,
            default: "",
            required: false
        },
        value: {
            type: Array,
            default() {
                return [];
            },
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: true,
            require: false
        }
    },


    watch: {
        // url
        url(){
            this.loadData(this.$props);
        }
    },

    mixins: [Transfer],

    methods: {
        loadData(params) {
            DataOp.fetchData(params, data => {
                this.dataSource = data;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
