// ui-jz-v4-common/src/components/jz-base/tree-select.js
/**
 * 此处填写文件描述
 * @Author: huangjq
 * @Date:   2019-04-02
 * @Project jz
 */

import { TreeSelect } from "ui-component-v4";
import DataOp from "../../utils/zt-data";
export default {
    name: TreeSelect.name,

    props: {
        data: {
            type: Array,
            default: () => [],
            required: false
        },
        // 表码
        code: {
            type: String,
            default: "",
            required: false
        },
        // 枚举
        enum: {
            type: String,
            default: "",
            required: false
        },
        // url
        url: {
            type: String,
            default: "",
            required: false
        },
        module: {
            type: String,
            default: "BASE_API"
        },
        // 不使用缓存
        noCache: {
            type: Boolean,
            default: false,
            required: false
        }
    },

    watch: {
        // 表码
        code() {
            this.loadData(this.$props);
        },
        // 枚举
        enum() {
            this.loadData(this.$props);
        },
        // url
        url(){
            this.loadData(this.$props);
        }
    },

    mixins: [TreeSelect],

    methods: {
        loadData(param) {
            DataOp.fetchData(param, data => {
                this.dataSource = data;
            });
        },
        initData() {
            this.loadData(this.$props);
        }
    },

    created() {
        this.initData();
    }
};
