// ui-jz-v4-common/src/components/jz-base/upload.js
/**
 * 文件上传
 * @Author: huangjq
 * @Date:   2019-03-04
 * @Project jz
 */

import { Upload } from "ui-component-v4";
const { getSetting } = window.systemParamsUtils;

export default {
    name: Upload.name,

    props: {
        // 上传的地址
        action: {
            type: String,
            default() {
                // /easy-mock/api/file/upload
                return `${getSetting("BASE_API")}/api/file/upload`;
            },
            required: false
        },

        // serverUrl: {
        //     type: String,
        //     default() {
        //         return `${getSetting("BASE_API")}`;
        //     },
        //     required: false
        // },

        fileList: {
            type: Array,
            default() {
                return [];
            }
        },

        headers: {
            type: Object,
            default() {
                let tokenObj = {};
                if (this.$store.getters.token){
                    tokenObj = { access_token: this.$store.getters.token }
                }
                return tokenObj;
            }
        },

        onSuccess: {
            type: Function,
            default: function(file, res) {
                res.id = file.id;
                res.url = file.url;
            },
            required: false
        },

        onExceed: {
            type: Function,
            default: function(files, fileList) {
                this.$message.warning(
                    `当前限制选择 ${this.limit} 个文件，本次选择了 ${files.length} 个文件，共选择了 ${files.length +
                        fileList.length} 个文件`
                );
            },
            required: false
        },
        resFunction: {
            type: Function,
            default: function(fileList) {
                return fileList; // .map(item => item.response);
            },
            required: false
        }
    },

    watch: {
        fileList(val) {
            return val;
        }
    },
    computed: {},

    methods: {
        handleSuccess(res, rawFile) {
            const file = this.getFile(rawFile);

            if (file) {
                file.status = "success";
                file.response = res;

                this.onSuccess(res, file, this.uploadFiles);
                this.onChange(file, this.uploadFiles);
            }
        }
    },

    mixins: [Upload],
    mounted() {},

    created() {}
};
