// ui-component-v4/packages/upload/src/unit.js
import _ from 'lodash';

// 文件单位
export const unit = {
    B: 1, // 1024^0
    K: 1024, // 1024^1
    M: 1048576, // 1024^2
    G: 1073741824, // 1024^3
    T: 1099511627776 // 1024^4
};

export function noop() {}

export function typeBySuffix(str = '') {
    const lastIdx = str.lastIndexOf('.');
    return lastIdx === -1 ? '' : str.slice(lastIdx + 1);
}

export function write2Class(type) {
    return !type ? 'el-icon-document' : `dg-icon-${type}`;
}

export function round(val) {
    return _.round(val, 2);
}

export function fileSize(size = 0) {
    if (size < unit.K) {
        return size + 'B';
    } else if (size < unit.M) {
        return round(size / unit.K) + 'KB';
    } else if (size < unit.G) {
        return round(size / unit.M) + 'MB';
    } else if (size < unit.T) {
        return round(size / unit.G) + 'GB';
    } else {
        return round(size / unit.T) + 'TB';
    }
}

export function sizeType(size) {
    return Object.prototype.toString.call(size);
}

// 下载
export function linkDown(url, fileName) {
    if (typeof url == 'object' && url instanceof Blob) {
        url = URL.createObjectURL(url);
    }

    let aLink = document.createElement('a');
    aLink.href = url;
    aLink.download = fileName || '';

    let event;

    if (window.MouseEvent) {
        event = new MouseEvent('click');
    } else {
        event = document.createEvent('MouseEvents');
        event.initMouseEvent('click', true, false, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
    }

    aLink.dispatchEvent(event);
}
