// ui-jz-v4-common/src/api/qb-base/abstract/AbsBufferedCodedicApi.js
import { BaseApi } from "../BaseApi";

/**
 * 代码表缓存
 * @description
 * @author
 * @date
 */
class AbsBufferedCodedicApi extends BaseApi {
    constructor(module) {
        super(module);
    }

    /**
     * 获取代码表
     * @param codeId
     * @param callback
     */
    getCodeMap(codeId,loading, callback) {
        throw new Error("未实现...");
    }
    /**
     * 通过url获取数据
     * @param url
     */
    fetchUrl(url) {
        throw new Error("未实现...");
    }

    /**
     * 获取某代码表具体编号的名称
     * @param codeId
     * @param bh
     * @param callback
     */
    getMc(codeId, bh,loading, callback) {
        throw new Error("未实现...");
    }
    /**
     * 刷新所有表码缓存
     */
    refreshAll(loading,callback) {
        throw new Error("未实现...");
    }
    /**
     * 根据名称获取编号
     * @param codeid
     * @param mc 表码名称
     * @return  表码编号
     */
    getBH(codeId, mc,loading, callback) {
        throw new Error("未实现...");
    }
    /**
     * 根据逗号分隔的编号获取名称
     * @param codeId
     * @param vBhs 逗号分隔字符串
     * @param callback
     */
    getMCS(codeId, vBhs,loading, callback) {
        throw new Error("未实现...");
    }
    /**
     * 刷新指定codeid的表码
     * @param codeId
     * @param callback
     */
    refreshCodeId(codeId, loading, callback) {
        throw new Error("未实现...");
    }
}
export { AbsBufferedCodedicApi };
