// ui-jz-v4-common/src/router/modules/about.js
/**
 * 框架介绍路由
 * @Author: huangjq
 * @Date:   2019-04-25
 * @Project jz
 */
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");

const componentsRouter = {
    path: "/about",
    component: Layout,
    meta: {
        title: "框架规范说明",
        icon: "el-icon-document"
    },
    redirect: "/about/dev-doc",
    children: [
        {
            path: "dev-doc",
            component: () => import("@/pages/about/dev-doc"),
            name: "devDoc",
            meta: { title: "开发相关", icon: "el-icon-document", noCache: false }
        }
        // {
        //     path: "init",
        //     component: () => import("@/pages/about/init"),
        //     name: "init",
        //     meta: { title: "快速搭建", icon: "dashboard", noCache: false }
        // },
        // {
        //     path: "quick-dev",
        //     component: () => import("@/pages/about/quick-dev"),
        //     name: "quickDev",
        //     meta: { title: "快速开发", icon: "dashboard", noCache: false }
        // },
        // {
        //     path: "dev-setting",
        //     component: () => import("@/pages/about/dev-setting"),
        //     name: "devSetting",
        //     meta: { title: "开发规范及配置", icon: "dashboard", noCache: false }
        // },
        // {
        //     path: "dockerDeploy",
        //     component: () => import("@/pages/about/dockerdeploy"),
        //     name: "dockerDeploy",
        //     meta: { title: "快速打包", icon: "dashboard", noCache: false }
        // },
    ]
};

export default componentsRouter;
