// ui-jz-v4-common/src/router/modules/components.js
/**
 * 组件demo路由
 * @Author: huangjq
 * @Date:   2019-04-11
 * @Project jz
 */
// import Layout from "@/pages/layout/layout";
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");
const componentsBasicPath = "pages/demo/components";

const componentsRouter = {
    path: "/components",
    component: Layout,
    meta: {
        title: "组件",
        icon: "el-icon-s-platform"
    },
    alwaysShow: true, // will always show the root menu
    children: [
        {
            path: "basic",
            component: () => import(`@/pages/common`),
            name: "BasicElements",
            redirect: "/components/basic/pc-alert",
            meta: { title: "基础元素", icon: "dashboard", noCache: false },
            children: [
                {
                    path: "pc-alert",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/alert`),
                    name: "pc-alert",
                    meta: { title: "警告", icon: "dashboard", noCache: false }
                },
                {
                    path: "pc-tag",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/tag`),
                    name: "pc-tag",
                    meta: { title: "标签", icon: "dashboard", noCache: false }
                },
                {
                    path: "pc-switch",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/switch`),
                    name: "pc-switch",
                    meta: { title: "开关", icon: "dashboard", noCache: false }
                },
                {
                    path: "pc-button",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/button`),
                    name: "pc-button",
                    meta: { title: "按钮", icon: "dashboard", noCache: false }
                },
                {
                    path: "dialog",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/dialog`),
                    name: "dialog",
                    meta: { title: "对话框", icon: "dashboard", noCache: true }
                },
                {
                    path: "message-box",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/message-box`),
                    name: "message-box",
                    meta: { title: "消息框", icon: "dashboard", noCache: true }
                },
                {
                    path: "scroll-bar",
                    component: () => import(`@/${componentsBasicPath}/basic-elements/scroll-bar`),
                    name: "scrollBar",
                    meta: { title: "滚动条", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "form-elements",
            component: () => import(`@/pages/common`),
            name: "formElements",
            redirect: "/components/form-elements/pc-input",
            meta: { title: "表单元素", icon: "dashboard", noCache: false },
            children: [
                {
                    path: "pc-input",
                    component: () => import(`@/${componentsBasicPath}/form-elements/input`),
                    name: "pc-input",
                    meta: { title: "输入框", icon: "dashboard", noCache: false }
                },
                {
                    path: "radio",
                    component: () => import(`@/${componentsBasicPath}/form-elements/radio`),
                    name: "Radio",
                    meta: { title: "单选框", icon: "dashboard", noCache: false }
                },
                {
                    path: "radioGroup",
                    component: () => import(`@/${componentsBasicPath}/form-elements/radio-group`),
                    name: "radioGroup",
                    meta: { title: "单选框组合", icon: "dashboard", noCache: false }
                },
                {
                    path: "checkGroup",
                    component: () => import(`@/${componentsBasicPath}/form-elements/checkbox-group`),
                    name: "checkGroup",
                    meta: { title: "复选框", icon: "dashboard", noCache: true }
                },
                {
                    path: "select",
                    component: () => import(`@/${componentsBasicPath}/form-elements/select`),
                    name: "select",
                    meta: { title: "下拉框", icon: "dashboard", noCache: true }
                },
                {
                    path: "transfer",
                    component: () => import(`@/${componentsBasicPath}/form-elements/transfer`),
                    name: "transfer",
                    meta: { title: "穿梭框", icon: "dashboard", noCache: true }
                },
                {
                    path: "cascader",
                    component: () => import(`@/${componentsBasicPath}/form-elements/cascader`),
                    name: "cascader",
                    meta: { title: "级联框", icon: "dashboard", noCache: true }
                },
                {
                    path: "upload",
                    component: () => import(`@/${componentsBasicPath}/form-elements/upload`),
                    name: "upload",
                    meta: { title: "文件上传", icon: "dashboard", noCache: true }
                },
                {
                    path: "treeDrop",
                    component: () => import(`@/${componentsBasicPath}/form-elements/tree-drop`),
                    name: "treeDrop",
                    meta: { title: "下拉树选择器", icon: "dashboard", noCache: true }
                },
                {
                    path: "tableSelect",
                    component: () => import(`@/${componentsBasicPath}/form-elements/table-select`),
                    name: "tableSelect",
                    meta: { title: "下拉表格选择", icon: "dashboard", noCache: true }
                },
                {
                    path: "date-picker",
                    component: () => import(`@/${componentsBasicPath}/form-elements/date-picker`),
                    name: "datePicker",
                    meta: { title: "日期选择器", icon: "dashboard", noCache: true }
                },
                {
                    path: "time-picker",
                    component: () => import(`@/${componentsBasicPath}/form-elements/time-picker`),
                    name: "timePicker",
                    meta: { title: "时间选择器", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "tables",
            component: () => import(`@/pages/common`),
            name: "tables",
            meta: { title: "表格", icon: "dashboard", noCache: false },
            children: [
                {
                    path: "table",
                    component: () => import(`@/${componentsBasicPath}/tables/table`),
                    name: "table",
                    meta: { title: "表格", icon: "dashboard", noCache: true }
                },
                {
                    path: "form-table",
                    component: () => import(`@/${componentsBasicPath}/tables/form-table`),
                    name: "form-table",
                    meta: { title: "表单表格", icon: "dashboard", noCache: true }
                },
                {
                    path: "line-table",
                    component: () => import(`@/${componentsBasicPath}/tables/line-table`),
                    name: "line-table",
                    meta: { title: "单行编辑表格", icon: "dashboard", noCache: true }
                },
                {
                    path: "operator",
                    component: () => import(`@/${componentsBasicPath}/operator`),
                    name: "operator",
                    meta: { title: "表单操作符", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "tree",
            component: () => import(`@/${componentsBasicPath}/tree`),
            name: "tree",
            meta: { title: "树形组件", icon: "dashboard", noCache: true }
        },
        {
            path: "excel",
            component: () => import(`@/${componentsBasicPath}/excel`),
            name: "excel",
            meta: { title: "excel导入(前端)", icon: "dashboard", noCache: true }
        },
        {
            path: "excel-import",
            component: () => import(`@/${componentsBasicPath}/excel-import`),
            name: "excel-import",
            meta: { title: "excel导入(后端)", icon: "dashboard", noCache: true }
        },
        {
            path: "ueditorWrap",
            component: () => import(`@/${componentsBasicPath}/ueditor`),
            name: "ueditorWrap",
            meta: { title: "富文本编辑器", icon: "dashboard", noCache: true }
        },
        {
            path: "viewer",
            component: () => import(`@/${componentsBasicPath}/viewer`),
            name: "viewer",
            meta: { title: "图片预览", icon: "dashboard", noCache: true }
        },
        {
            path: "chart",
            component: () => import(`@/${componentsBasicPath}/chart`),
            name: "chart",
            meta: { title: "图表", icon: "dashboard", noCache: true }
        },
        {
            path: "timeLine",
            component: () => import(`@/${componentsBasicPath}/time-line`),
            name: "TimeLine",
            meta: { title: "时间轴", icon: "dashboard", noCache: true }
        },
        {
            path: "logSearch",
            component: () => import(`@/${componentsBasicPath}/log-search`),
            name: "LogSearch",
            meta: { title: "日志查询", icon: "dashboard", noCache: true }
        },
        // {
        //     path: "audio",
        //     component: () => import(`@/${componentsBasicPath}/audio`),
        //     name: "Audio",
        //     meta: { title: "音乐播放", icon: "dashboard", noCache: true }
        // },
        // {
        //     path: "video",
        //     component: () => import(`@/${componentsBasicPath}/video`),
        //     name: "Video",
        //     meta: { title: "视频播放", icon: "dashboard", noCache: true }
        // },
        // {
        //     path: "codemirror",
        //     component: () => import(`@/${componentsBasicPath}/codemirror`),
        //     name: "Codemirror",
        //     meta: { title: "代码编辑器", icon: "dashboard", noCache: true }
        // },
        {
            path: "pageCode",
            component: () => import(`@/${componentsBasicPath}/page-code`),
            name: "PageCode",
            meta: { title: "状态码", icon: "dashboard", noCache: true }
        },
        {
            path: "card",
            component: () => import(`@/${componentsBasicPath}/card`),
            name: "Card",
            meta: { title: "卡片", icon: "dashboard", noCache: true }
        },
        {
            path: "badge",
            component: () => import(`@/${componentsBasicPath}/badge`),
            name: "Badge",
            meta: { title: "标记", icon: "dashboard", noCache: true }
        },
        {
            path: "message",
            component: () => import(`@/${componentsBasicPath}/message`),
            name: "Message",
            meta: { title: "消息提示", icon: "dashboard", noCache: true }
        },
        {
            path: "notification",
            component: () => import(`@/${componentsBasicPath}/notification`),
            name: "Notification",
            meta: { title: "通知", icon: "dashboard", noCache: true }
        }
    ]
};

export default componentsRouter;
