// ui-jz-v4-common/src/router/modules/demos.js
/**
 * 框架基础案例
 * @Author: huangjq
 * @Date:   2020-04-20
 * @Project jz
 */
// import Layout from "@/pages/layout/layout";
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");

const componentsRouter = {
    path: "/demos",
    component: Layout,
    redirect: "/demos/login-page",
    meta: { title: "案例", icon: "el-icon-mobile-phone", noCache: true },
    children: [
        

        {
            path: "login-page",
            component: () => import(`@/pages/common`),
            meta: {
                title: "登录",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "login-type-1",
                    component: () => import("@/pages/demo/login-page/login-type-1.vue"),
                    name: "login-type-1",
                    meta: {
                        title: "登录",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        },
        {
            path: "space",
            component: () => import(`@/pages/common`),
            meta: {
                title: "间距页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "spaceLayout",
                    component: () => import("@/pages/demo/space/space-layout.vue"),
                    name: "spaceLayout",
                    meta: {
                        title: "基础间距布局",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "spaceHeaderLayout",
                    component: () => import("@/pages/demo/space/space-header-layout.vue"),
                    name: "spaceHeaderLayout",
                    meta: {
                        title: "标题间距布局",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "spaceBaseLayout",
                    component: () => import("@/pages/demo/space/space-base-layout.vue"),
                    name: "spaceBaseLayout",
                    meta: {
                        title: "边框间距布局",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        },
        {
            path: "demoForm",
            component: () => import(`@/pages/common`),
            meta: {
                title: "表单页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "formSingle",
                    component: () => import("@/pages/demo/demo-form/form-single.vue"),
                    name: "formSingle",
                    meta: {
                        title: "单列表单页",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "formDouble",
                    component: () => import("@/pages/demo/demo-form/form-double.vue"),
                    name: "formDouble",
                    meta: {
                        title: "双列表单页",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "distribution-form",
                    name: "分步表单",
                    component: () => import("@/pages/demo/distribution-form/index"),
                    meta: {
                        title: "分步表单",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                },
                {
                    path: "zfba",
                    name: "三列表单",
                    component: () => import("@/pages/demo/zfba/case-acceptance/index.vue"),
                    meta: {
                        title: "三列表单",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                }
            ]
        },
        {
            path: "list",
            component: () => import(`@/pages/common`),
            meta: {
                title: "列表页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "search-list-1",
                    name: "search-list-1",
                    component: () => import("@/pages/demo/list/search-list-1/index.vue"),
                    meta: {
                        title: "搜索列表-卡片1",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                },
                {
                    path: "search-list-2",
                    name: "search-list-2",
                    component: () => import("@/pages/demo/list/search-list-2/index.vue"),
                    meta: {
                        title: "搜索列表-卡片2",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                },
                {
                    path: "search-list-3",
                    name: "search-list-3",
                    component: () => import("@/pages/demo/list/search-list-3/index.vue"),
                    meta: {
                        title: "搜索列表-卡片3",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                },
                {
                    path: "search-list-4",
                    name: "search-list-4",
                    component: () => import("@/pages/demo/list/search-list-4/index.vue"),
                    meta: {
                        title: "搜索列表-卡片4",
                        icon: "dashboard",
                        layout: "page",
                        noCache: false
                    }
                }
            ]
        },
        {
            path: "demoTable",
            component: () => import(`@/pages/common`),
            meta: {
                title: "表格页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "tableToggle",
                    component: () => import("@/pages/demo/demo-table/table-toggle.vue"),
                    name: "tableToggle",
                    meta: {
                        title: "查询表格",
                        icon: "dashboard",
                        noCache: false
                    }
                }
            ]
        },
        {
            path: "demoDialog",
            component: () => import(`@/pages/common`),
            meta: {
                title: "弹窗页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "dialogIndex",
                    component: () => import("@/pages/demo/demo-dialog/index.vue"),
                    name: "dialogIndex",
                    meta: {
                        title: "弹窗页",
                        icon: "dashboard",
                        noCache: false
                    }
                }
            ]
        },
        {
            path: "result-page",
            component: () => import(`@/pages/common`),
            meta: {
                title: "结果页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "submit-successfully",
                    component: () => import("@/pages/demo/result-page/submit-successfully.vue"),
                    name: "submit-successfully",
                    meta: {
                        title: "提交成功",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "submit-failure",
                    component: () => import("@/pages/demo/result-page/submit-failure.vue"),
                    name: "submit-failure",
                    meta: {
                        title: "提交失败",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        },
        {
            path: "exception",
            component: () => import(`@/pages/common`),
            meta: {
                title: "异常页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "403",
                    component: () => import("@/pages/demo/exception/403.vue"),
                    name: "403",
                    meta: {
                        title: "403 页面",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "404",
                    component: () => import("@/pages/demo/exception/404.vue"),
                    name: "404",
                    meta: {
                        title: "404 页面",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "500",
                    component: () => import("@/pages/demo/exception/500.vue"),
                    name: "500",
                    meta: {
                        title: "500 页面",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "501",
                    component: () => import("@/pages/demo/exception/501.vue"),
                    name: "501",
                    meta: {
                        title: "501 页面",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        },
        {
            path: "details-page",
            component: () => import(`@/pages/common`),
            meta: {
                title: "详情页",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "basics-page",
                    component: () => import("@/pages/demo/details-page/basics-page.vue"),
                    name: "basics-page",
                    meta: {
                        title: "基础详情页",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "list-details-page",
                    component: () => import("@/pages/demo/details-page/list-details-page.vue"),
                    name: "list-details-page",
                    meta: {
                        title: "列表详情页",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "senior-page",
                    component: () => import("@/pages/demo/details-page/senior-page.vue"),
                    name: "senior-page",
                    meta: {
                        title: "高级详情页",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        },
        {
            path: "demoSubfield",
            component: () => import(`@/pages/common`),
            meta: {
                title: "三级侧边栏",
                icon: "dashboard",
                noCache: false
            },
            children: [
                {
                    path: "subfieldOutside",
                    component: () => import("@/pages/demo/demo-subfield/subfield-outside.vue"),
                    name: "subfieldOutside",
                    meta: {
                        title: "分栏外侧布局",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                },
                {
                    path: "subfieldInside",
                    component: () => import("@/pages/demo/demo-subfield/subfield-inside.vue"),
                    name: "subfieldInside",
                    meta: {
                        title: "分栏内侧布局",
                        icon: "dashboard",
                        noCache: false,
                        layout: "page"
                    }
                }
            ]
        }
    
    ]
};

export default componentsRouter;
