// ui-jz-v4-common/src/router/modules/demo.js
/**
 * 实例路由
 * @Author: huangjq
 * @Date:   2019-04-11
 * @Project jz
 */
// import Layout from "@/pages/layout/layout";
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");
import "!raw-loader";
import Vue from "vue";
const fileArr = [
    {
        key: "simple-manager-edit",
        value: "simple-manager/edit.htm",
        isEdit: true
    },
    {
        key: "simple-manager",
        value: "simple-manager/list.htm"
    },
    {
        key: "org-manager-edit",
        value: "org-manager/edit.htm",
        isEdit: true
    },
    {
        key: "org-manager",
        value: "org-manager/list.htm"
    },
    {
        key: "dialog-demo",
        value: "components/basic-elements/dialog/index.htm",
        isEdit: true
    },
    {
        key: "async-page",
        value: "async-fields/index.htm"
    }
];
//
Vue.mixin({
    mounted() {
        let fileObj = "";
        let isCatch = fileArr.some(item => {
            fileObj = item;
            return this.$vnode && this.$vnode.tag.indexOf(item.key) > 0;
        });
        if (this.$vnode && isCatch) {
            let $btn = $(
                "<button style='position: absolute;z-index:1900;top: 0px;right: 16px;' class='el-button el-button--primary'>查看源码</button>"
            );
            let $el = this.$el;
            if (fileObj.isEdit) {
                $btn.css("top", "16px");
                $btn.css("z-index", "1900");
            }
            $btn.prependTo($($el));
            $btn.on("click", () => {
                let path = `${fileObj.value}`;
                let code = require("@/pages/demo/" + path);
                this.$dgLayer({
                    title: "源码",
                    content: `<textarea style="width: 99%;height: 97%;border: none;">${code}</textarea>`,
                    type: 1,
                    props: {},
                    on: {
                        success(result) {
                            console.log(result)
                        }
                    },
                    cancel: function(index, layero) {
                        console.log("invoke cancel", index, layero);
                        // return false;
                    },
                    area: ["60%", "80%"]
                });
            });
        }
    }
});
const demosRouter = {
    path: "/demo",
    component: Layout,
    redirect: "/demo/simpleManager",
    meta: {
        title: "demo",
        icon: "icon-file",
        noCache: false
    },
    children: [
        {
            path: "simpleManager",
            component: () => import("@/pages/demo/simple-manager/list.vue"),
            name: "simple-manager",
            meta: {
                title: "简单增删改查客户端分页实例",
                icon: "dashboard",
                noCache: false
            }
        },
        {
            path: "org-manager",
            component: () => import("@/pages/demo/org-manager/list.vue"),
            name: "org-manager",
            meta: {
                title: "组织机构增删改查实例",
                icon: "dashboard",
                noCache: false
            }
        }
        // {
        //     path: "asyncPage",
        //     component: () => import("@/pages/demo/async-fields"),
        //     name: "asyncPage",
        //     meta: {
        //         title: "动态数据",
        //         icon: "icon-file",
        //         noCache: false
        //     }
        // },
        // {
        //     path: "docking8081",
        //     component: () => import("@/pages/demo/docking/index.vue"),
        //     name: "Docking8081",
        //     meta: {
        //         title: "对接页面8081",
        //         icon: "icon-file",
        //         noCache: true
        //     }
        // },
        // {
        //     path: "goJs",
        //     component: () => import("@/pages/demo/go/workflow.vue"),
        //     name: "goJs",
        //     meta: {
        //         title: "拓扑图",
        //         icon: "dashboard",
        //         noCache: true
        //     }
        // },
    ]
};

export default demosRouter;
