// ui-jz-v4-common/src/router/modules/mobile.js
/**
 * 框架介绍路由
 * @Author: huangjq
 * @Date:   2019-04-25
 * @Project jz
 */
// import Layout from "@/pages/layout/layout";
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");

const componentsRouter = {
    path: "/mobile",
    component: Layout,
    redirect: "/mobile/about",
    meta: { title: "手机相关", icon: "el-icon-mobile-phone", noCache: true },
    children: [
        {
            path: "about",
            component: () => import("@/pages/about/mobile/about"),
            name: "about",
            meta: { title: "环境搭建", icon: "dashboard", noCache: true }
        },
        {
            path: "guide",
            component: () => import("@/pages/about/mobile/guide"),
            name: "guide",
            meta: { title: "框架简介", icon: "dashboard", noCache: true }
        },
        {
            path: "component",
            component: () => import("@/pages/about/mobile/components"),
            name: "component",
            meta: { title: "组件", icon: "dashboard", noCache: true },
            redirect: "/mobile/component/input",
            children: [
                {
                    path: "input",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "input",
                    meta: { title: "input", icon: "dashboard", noCache: true }
                },
                {
                    path: "textarea",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "textarea",
                    meta: { title: "textarea", icon: "dashboard", noCache: true }
                },
                {
                    path: "image",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "image",
                    meta: { title: "image", icon: "dashboard", noCache: true }
                },
                {
                    path: "button",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "button",
                    meta: { title: "button", icon: "dashboard", noCache: true }
                },
                {
                    path: "checkbox",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "checkbox",
                    meta: { title: "checkbox", icon: "dashboard", noCache: true }
                },
                {
                    path: "radio",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "radio",
                    meta: { title: "radio", icon: "dashboard", noCache: true }
                },
                {
                    path: "progress",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "progress",
                    meta: { title: "progress", icon: "dashboard", noCache: true }
                },
                {
                    path: "switch",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "switch",
                    meta: { title: "switch", icon: "dashboard", noCache: true }
                },
                {
                    path: "slider",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "slider",
                    meta: { title: "slider", icon: "dashboard", noCache: true }
                },
                {
                    path: "swiper",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "swiper",
                    meta: { title: "swiper", icon: "dashboard", noCache: true }
                },
                {
                    path: "scroll-view",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "scroll-view",
                    meta: { title: "scroll-view", icon: "dashboard", noCache: true }
                },
                {
                    path: "text",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "text",
                    meta: { title: "text", icon: "dashboard", noCache: true }
                },
                {
                    path: "audio",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "audio",
                    meta: { title: "audio", icon: "dashboard", noCache: true }
                },
                {
                    path: "video",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "video",
                    meta: { title: "video", icon: "dashboard", noCache: true }
                },
                {
                    path: "web-view",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "web-view",
                    meta: { title: "web-view", icon: "dashboard", noCache: true }
                },
                {
                    path: "navigator",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "navigator",
                    meta: { title: "navigator", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "API",
            component: () => import("@/pages/about/mobile/components"),
            name: "API",
            meta: { title: "接口", icon: "dashboard", noCache: true },
            redirect: "/mobile/component/log",
            children: [
                {
                    path: "log",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "log",
                    meta: { title: "日志打印", icon: "dashboard", noCache: true }
                },
                {
                    path: "request",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "request",
                    meta: { title: "网络请求", icon: "dashboard", noCache: true }
                },
                {
                    path: "network-file",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "network-file",
                    meta: { title: "上传下载", icon: "dashboard", noCache: true }
                },
                {
                    path: "file",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "file",
                    meta: { title: "文件存储", icon: "dashboard", noCache: true }
                },
                {
                    path: "storage",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "storage",
                    meta: { title: "数据缓存", icon: "dashboard", noCache: true }
                },
                {
                    path: "api-image",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "api-image",
                    meta: { title: "媒体图片", icon: "dashboard", noCache: true }
                },
                {
                    path: "api-video",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "api-video",
                    meta: { title: "媒体视频", icon: "dashboard", noCache: true }
                },
                {
                    path: "location",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "location",
                    meta: { title: "位置定位", icon: "dashboard", noCache: true }
                },
                {
                    path: "network",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "network",
                    meta: { title: "网络状态", icon: "dashboard", noCache: true }
                },
                {
                    path: "phone",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "phone",
                    meta: { title: "拨打电话", icon: "dashboard", noCache: true }
                },
                {
                    path: "prompt",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "prompt",
                    meta: { title: "交互对话", icon: "dashboard", noCache: true }
                },
                {
                    path: "navigationbar",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "navigationbar",
                    meta: { title: "设置导航栏", icon: "dashboard", noCache: true }
                },
                {
                    path: "tabbar",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "tabbar",
                    meta: { title: "设置TABBAR", icon: "dashboard", noCache: true }
                },
                {
                    path: "pulldown",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "pulldown",
                    meta: { title: "下拉刷新", icon: "dashboard", noCache: true }
                },
                {
                    path: "router",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "router",
                    meta: { title: "路由", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "collocation",
            component: () => import("@/pages/about/mobile/components"),
            name: "collocation",
            meta: { title: "配置", icon: "dashboard", noCache: true },
            redirect: "/mobile/component/page",
            children: [
                {
                    path: "page",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "page",
                    meta: { title: "pages.json", icon: "dashboard", noCache: true }
                },
                {
                    path: "manifest",
                    component: () => import("@/pages/about/mobile/components"),
                    name: "manifest",
                    meta: { title: "manifest.json", icon: "dashboard", noCache: true }
                }
            ]
        }
    ]
};

export default componentsRouter;
