// ui-jz-v4-common/src/router/modules/qb-components.js
/**
 *@author:
 *@createTime: 2019/10/16 15:34
 *@description:qb组件路由
 *@功能:
 */
// import Layout from "@/pages/layout/layout";
const Layout = () => import("@/pages/" + process.env.LAYOUT + "/layout");
const componentsBasicPath = "pages/demo/qb-components";

const componentsRouter = {
    path: "/qb-components",
    component: Layout,
    meta: {
        title: "qb组件",
        icon: "el-icon-s-platform" 
    },
    alwaysShow: true, // will always show the root menu
    children: [
        {
            path: "qb-form-elements",
            component: () => import(`@/pages/common`),
            name: "qbFormElements",
            meta: { title: "表单元素", icon: "dashboard", noCache: false },
            children: [
                {
                    path: "qbShowLable",
                    component: () => import(`@/${componentsBasicPath}/form-elements/show-lable`),
                    name: "qbShowLable",
                    meta: { title: "showLable", icon: "dashboard", noCache: false }
                },
                {
                    path: "qbRadioGroup",
                    component: () => import(`@/${componentsBasicPath}/form-elements/radio-group`),
                    name: "qbRadioGroup",
                    meta: { title: "单选框组合", icon: "dashboard", noCache: false }
                },
                {
                    path: "qbCheckGroup",
                    component: () => import(`@/${componentsBasicPath}/form-elements/checkbox-group`),
                    name: "qbCheckGroup",
                    meta: { title: "复选框", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbSelect",
                    component: () => import(`@/${componentsBasicPath}/form-elements/select`),
                    name: "qbSelect",
                    meta: { title: "下拉框", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbTransfer",
                    component: () => import(`@/${componentsBasicPath}/form-elements/transfer`),
                    name: "qbTransfer",
                    meta: { title: "穿梭框", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbCascader",
                    component: () => import(`@/${componentsBasicPath}/form-elements/cascader`),
                    name: "qbCascader",
                    meta: { title: "级联框", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbUpload",
                    component: () => import(`@/${componentsBasicPath}/form-elements/upload`),
                    name: "qbUpload",
                    meta: { title: "文件上传", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbTreeSelect",
                    component: () => import(`@/${componentsBasicPath}/form-elements/tree-select`),
                    name: "qbTreeSelect",
                    meta: { title: "下拉树选择器", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbTableSelect",
                    component: () => import(`@/${componentsBasicPath}/form-elements/table-select`),
                    name: "qbTableSelect",
                    meta: { title: "下拉表格选择", icon: "dashboard", noCache: true }
                },
                {
                    path: "qbGroupSelect",
                    component: () => import(`@/${componentsBasicPath}/form-elements/group-select`),
                    name: "qbGroupSelect",
                    meta: { title: "分组选择器", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "qbTables",
            component: () => import(`@/pages/common`),
            name: "qbTables",
            meta: { title: "表格", icon: "dashboard", noCache: false },
            children: [
                {
                    path: "qbTable",
                    component: () => import(`@/${componentsBasicPath}/tables/table`),
                    name: "qbTable",
                    meta: { title: "表格", icon: "dashboard", noCache: true }
                }
            ]
        },
        {
            path: "qbTree",
            component: () => import(`@/${componentsBasicPath}/tree`),
            name: "qbTree",
            meta: { title: "树形组件", icon: "dashboard", noCache: true }
        },
        {
            path: "qbTimeLine",
            component: () => import(`@/${componentsBasicPath}/time-line`),
            name: "qbTimeLine",
            meta: { title: "时间轴", icon: "dashboard", noCache: true }
        }
    ]
};

export default componentsRouter;
